{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotate complexes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\" id=\"colomoto-setup-5204324\" class=\"to-be-removed\">\n",
       "        if (typeof Jupyter != 'undefined') {\n",
       "            \n",
       "function detect_import(cell, module) {\n",
       "    var code = cell.get_text();\n",
       "    code = code.replace(/\\\\\\n/g, \"\");\n",
       "    var lines = code.split(\"\\n\");\n",
       "    var r_simple = new RegExp(\"^(\"+module+\")$\");\n",
       "    var r_alias = new RegExp(\"^\"+module+\"\\\\s+as\\\\s+(\\\\w+)$\");\n",
       "    for (var i = 0; i < lines.length; ++i) {\n",
       "        if (/^import\\s/.test(lines[i])) {\n",
       "            code = lines[i].substr(7);\n",
       "            var parts = code.split(\",\")\n",
       "            for (var j = 0; j < parts.length; ++j) {\n",
       "                code = parts[j].trim();\n",
       "                var m = code.match(r_simple);\n",
       "                if (!m) {\n",
       "                    m = code.match(r_alias);\n",
       "                }\n",
       "                if (m) {\n",
       "                    return m[1];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    return module;\n",
       "}\n",
       "\n",
       "function colomoto_replace_call(cell, orig, dest, args, comment=false) {\n",
       "    var call_regexp = new RegExp(\"\\\\.\"+orig.replace(\".\",\"\\\\.\")+\"\\\\(\");\n",
       "    var call_replacer = new RegExp(\"\\\\.\"+orig.replace(\".\",\"\\\\.\")\n",
       "                    + \"\\\\(\\\\s*([^\\\\)]*)?\\\\)\");\n",
       "    var code = cell.get_text();\n",
       "    var lines = code.split(\"\\n\");\n",
       "    if (args) {\n",
       "        var strargs = \", \"+args.join(\", \");\n",
       "    } else {\n",
       "        var strargs = \"\";\n",
       "    }\n",
       "    for (var i = 0; i < lines.length; ++i) {\n",
       "        if (call_regexp.test(lines[i])) {\n",
       "            var code = \"\"\n",
       "            if (comment) {\n",
       "                code += \"#\"+lines[i]+\"\\n\";\n",
       "            }\n",
       "            code += lines[i].replace(call_replacer, \".\"+dest+\"($1\"+strargs+\")\")\n",
       "            lines[i] = code;\n",
       "        }\n",
       "    }\n",
       "    cell.set_text(code)\n",
       "}\n",
       "\n",
       "function colomoto_upload(Jupyter, ssid, input, py_callback_name, orig, dest) {\n",
       "\n",
       "    function callback(out_data) {\n",
       "        var cell_element = $(\"#\"+ssid).parents('.cell');\n",
       "        var cell_idx = Jupyter.notebook.get_cell_elements().index(cell_element);\n",
       "        var cell = Jupyter.notebook.get_cell(cell_idx);\n",
       "\n",
       "        var filename = out_data.content.text;\n",
       "\n",
       "        var code = cell.get_text();\n",
       "        code = code.replace(new RegExp(\"\\\\b\" + orig.replace('.', '\\\\.')\n",
       "                    + \"\\\\(\\\\s*((\\\\w+)=[^\\\\)]*)?\\\\)\"),\n",
       "                dest+\"(\\\"\"+filename+\"\\\",$1)\");\n",
       "        code = code.replace('\",)', '\")')\n",
       "        cell.set_text(code);\n",
       "\n",
       "        Jupyter.notebook.select(cell_idx);\n",
       "        Jupyter.notebook.execute_cell_and_select_below();\n",
       "    }\n",
       "\n",
       "    if (! (window.File && window.FileReader && window.FileList && window.Blob)) {\n",
       "        alert(\"Interactive file upload is not supported by your browser.\");\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    input.disabled = true;\n",
       "    input.style.cursor = \"wait\";\n",
       "    input.parentElement.style.cursor = \"wait\";\n",
       "\n",
       "    var f = input.files[0];\n",
       "    var reader = new FileReader();\n",
       "    reader.onload = (function(f) {\n",
       "        return function (e) {\n",
       "            var obj = {\n",
       "                content: e.target.result,\n",
       "                name: f.name\n",
       "            };\n",
       "\n",
       "            //var pycb = py_callback_name+\"(\"+JSON.stringify(obj)+\")\"\n",
       "            // hack/workaround:\n",
       "            // it seems that Jupyter does not like very long lines\n",
       "            // so we split the data in chunks\n",
       "            var chunk_length = 100;\n",
       "            var pycb = \"__colomoto_upload_name = \" + JSON.stringify(obj.name) + \"\\n\";\n",
       "            pycb += \"__colomoto_upload_content = \\\\\\n\";\n",
       "            for (var i = 0; i < obj.content.length; i += chunk_length) {\n",
       "                pycb += \"\\\"\" + obj.content.substr(i, chunk_length)+\"\\\"\\\\\\n\"\n",
       "            }\n",
       "            pycb += \"\\n\"\n",
       "            pycb += py_callback_name+\"({'name':__colomoto_upload_name, 'content': __colomoto_upload_content})\\n\";\n",
       "            pycb += \"del __colomoto_upload_name, __colomoto_upload_content\"\n",
       "\n",
       "            IPython.notebook.kernel.execute(pycb, {iopub: {output: callback}});\n",
       "        };\n",
       "    })(f);\n",
       "    reader.readAsDataURL(f);\n",
       "}\n",
       "\n",
       "function resolve_function(tool_api, funcname) {\n",
       "    if (tool_api.hasOwnProperty(funcname)) {\n",
       "        return tool_api[funcname];\n",
       "    } else {\n",
       "        return window[funcname];\n",
       "    }\n",
       "}\n",
       "\n",
       "function colomoto_extension(Jupyter, ssid, name, menu, toolbar, tool_api) {\n",
       "\n",
       "    function insert_snippet_code(snippet) {\n",
       "        var cell = Jupyter.notebook.get_selected_cell();\n",
       "        Jupyter.notebook.edit_mode();\n",
       "        cell.code_mirror.replaceSelection(snippet, 'around');\n",
       "        //cell.focus_editor();\n",
       "    }\n",
       "\n",
       "    /**\n",
       "        from https://github.com/moble/jupyter_boilerplate/blob/master/main.js\n",
       "    */\n",
       "    function callback_insert_snippet (evt) {\n",
       "        // this (or event.currentTarget, see below) always refers to the DOM\n",
       "        // element the listener was attached to - see\n",
       "        // http://stackoverflow.com/questions/12077859\n",
       "        insert_snippet_code($(evt.currentTarget).data('snippet-code'));\n",
       "    }\n",
       "    function build_menu_element (menu_item_spec, direction) {\n",
       "        // Create the menu item html element\n",
       "        var element = $('<li/>');\n",
       "\n",
       "        if (typeof menu_item_spec == 'string') {\n",
       "            if (menu_item_spec != '---') {\n",
       "                return element.html(menu_item_spec)\n",
       "                        .addClass('ui-state-disabled')\n",
       "                        .attr({\"style\": \"padding:2px .4em\"})\n",
       "                       ;\n",
       "            }\n",
       "            return element.addClass('divider');\n",
       "        }\n",
       "\n",
       "        var a = $('<a/>')\n",
       "            .attr('href', '#')\n",
       "            .html(menu_item_spec.name)\n",
       "            .appendTo(element);\n",
       "        if (menu_item_spec.hasOwnProperty('snippet')) {\n",
       "            var snippet = menu_item_spec.snippet;\n",
       "            if (typeof snippet == 'string' || snippet instanceof String) {\n",
       "                snippet = [snippet];\n",
       "            }\n",
       "            a.attr({\n",
       "                'title' : \"\", // Do not remove this, even though it's empty!\n",
       "                'data-snippet-code' : snippet.join('\\n'),\n",
       "            })\n",
       "            .on('click', callback_insert_snippet)\n",
       "            .addClass('snippet');\n",
       "        }\n",
       "        else if (menu_item_spec.hasOwnProperty('internal-link')) {\n",
       "            a.attr('href', menu_item_spec['internal-link']);\n",
       "        }\n",
       "        else if (menu_item_spec.hasOwnProperty('external-link')) {\n",
       "            a.empty();\n",
       "            a.attr('href', menu_item_spec['external-link']);\n",
       "            a.attr({\n",
       "                'target' : '_blank',\n",
       "                'title' : 'Opens in a new window',\n",
       "            });\n",
       "            $('<i class=\"fa fa-external-link menu-icon pull-right\"/>').appendTo(a);\n",
       "            $('<span/>').html(menu_item_spec.name).appendTo(a);\n",
       "        }\n",
       "\n",
       "        if (menu_item_spec.hasOwnProperty('sub-menu')) {\n",
       "            element\n",
       "                .addClass('dropdown-submenu')\n",
       "                .toggleClass('dropdown-submenu-left', direction === 'left');\n",
       "            var sub_element = $('<ul class=\"dropdown-menu\"/>')\n",
       "                .toggleClass('dropdown-menu-compact', menu_item_spec.overlay === true) // For space-saving menus\n",
       "                .appendTo(element);\n",
       "\n",
       "            var new_direction = (menu_item_spec['sub-menu-direction'] === 'left') ? 'left' : 'right';\n",
       "            for (var j=0; j<menu_item_spec['sub-menu'].length; ++j) {\n",
       "                var sub_menu_item_spec = build_menu_element(menu_item_spec['sub-menu'][j], new_direction);\n",
       "                if(sub_menu_item_spec !== null) {\n",
       "                    sub_menu_item_spec.appendTo(sub_element);\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "\n",
       "        return element;\n",
       "    }\n",
       "\n",
       "    function menu_setup (menu_item_specs, sibling, insert_before_sibling) {\n",
       "        for (var i=0; i<menu_item_specs.length; ++i) {\n",
       "            var menu_item_spec;\n",
       "            if (insert_before_sibling) {\n",
       "                menu_item_spec = menu_item_specs[i];\n",
       "            } else {\n",
       "                menu_item_spec = menu_item_specs[menu_item_specs.length-1-i];\n",
       "            }\n",
       "            var direction = (menu_item_spec['menu-direction'] == 'left') ? 'left' : 'right';\n",
       "            var menu_element = build_menu_element(menu_item_spec, direction);\n",
       "            // We need special properties if this item is in the navbar\n",
       "            if ($(sibling).parent().is('ul.nav.navbar-nav')) {\n",
       "                menu_element\n",
       "                    .addClass('dropdown')\n",
       "                    .removeClass('dropdown-submenu dropdown-submenu-left');\n",
       "                menu_element.children('a')\n",
       "                    .addClass('dropdown-toggle')\n",
       "                    .attr({\n",
       "                        'id': name+'_menu',\n",
       "                        'data-toggle' : 'dropdown',\n",
       "                        'aria-expanded' : 'false'\n",
       "                    });\n",
       "            }\n",
       "\n",
       "            // Insert the menu element into DOM\n",
       "            menu_element[insert_before_sibling ? 'insertBefore': 'insertAfter'](sibling);\n",
       "        }\n",
       "    }\n",
       "    /** end from */\n",
       "\n",
       "\n",
       "    function self_cleanup() {\n",
       "        var cell_element = $(\"script[class='to-be-removed']\").parents('.cell');\n",
       "        var cell_idx = Jupyter.notebook.get_cell_elements().index(cell_element);\n",
       "        var cell = Jupyter.notebook.get_cell(cell_idx);\n",
       "        var to_remove = -1;\n",
       "        for (var i = 0; i < cell.output_area.outputs.length; ++i) {\n",
       "            var oa = cell.output_area.outputs[i];\n",
       "            if (oa.output_type == \"display_data\"\n",
       "                && typeof oa.data[\"text/html\"] != 'undefined'\n",
       "                && oa.data[\"text/html\"].indexOf(' class=\"to-be-removed\"') >= 0) {\n",
       "                to_remove = i;\n",
       "                break;\n",
       "            }\n",
       "        }\n",
       "        if (to_remove == -1) {\n",
       "            console.log(\"cannot find toberemoved\");\n",
       "        } else {\n",
       "            cell.output_area.outputs.splice(to_remove, 1);\n",
       "        }\n",
       "    }\n",
       "\n",
       "    function toolbar_setup(actions) {\n",
       "        var buttons = [];\n",
       "        for (var i = 0; i < actions.length; ++i) {\n",
       "            var setup = actions[i].setup;\n",
       "            if (typeof setup.handler == 'string') {\n",
       "                setup.handler = resolve_function(tool_api, setup.handler);\n",
       "            }\n",
       "            buttons.push(Jupyter.actions.register(actions[i].setup,\n",
       "                actions[i].name, name));\n",
       "        }\n",
       "        $(\"#\"+name+\"-toolbar\").remove();\n",
       "        Jupyter.toolbar.add_buttons_group(buttons, name+\"-toolbar\");\n",
       "    }\n",
       "\n",
       "    function replace_menu_snippets(menu_spec, orig, dest) {\n",
       "        if (menu_spec.hasOwnProperty(\"snippet\")) {\n",
       "            var snippet = menu_spec.snippet;\n",
       "            if (typeof snippet == \"string\" || snippet instanceof String) {\n",
       "                menu_spec[\"snippet\"] = snippet.replace(orig, dest);\n",
       "            } else {\n",
       "                for (var i = 0; i < snippet.length; ++i) {\n",
       "                    menu_spec[\"snippet\"][i] = snippet[i].replace(orig, dest);\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "        if (menu_spec.hasOwnProperty(\"sub-menu\")) {\n",
       "            for (var i = 0; i < menu_spec[\"sub-menu\"].length; ++i) {\n",
       "                replace_menu_snippets(menu_spec[\"sub-menu\"][i], orig, dest);\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "\n",
       "    function load_ipython_extension() {\n",
       "\n",
       "        var mycellelt = $(\"#\"+ssid).parents('.cell');\n",
       "        var myidx = Jupyter.notebook.get_cell_elements().index(mycellelt);\n",
       "        var import_cell = Jupyter.notebook.get_cell(myidx);\n",
       "\n",
       "        var alias = detect_import(import_cell, name);\n",
       "        tool_api.module_alias = alias;\n",
       "        if (alias && alias != name) {\n",
       "            var orig = new RegExp(\"\\\\b\"+name+\"\\\\b\", \"g\");\n",
       "            replace_menu_snippets(menu, orig, alias);\n",
       "        }\n",
       "\n",
       "        if (toolbar) {\n",
       "            toolbar_setup(toolbar);\n",
       "        }\n",
       "\n",
       "        $(\"#\"+name+\"_menu\").parent().remove();\n",
       "        if (menu) {\n",
       "            menu_setup([menu], $(\"#help_menu\").parent(), true);\n",
       "        }\n",
       "\n",
       "        if (tool_api.hasOwnProperty(\"post_install_callback\")) {\n",
       "            tool_api.post_install_callback();\n",
       "        }\n",
       "\n",
       "        setTimeout(self_cleanup, 5000);\n",
       "    };\n",
       "\n",
       "    load_ipython_extension();\n",
       "}\n",
       "\n",
       "function resolve_toolbar_handlers(tool_api, toolbar_spec) {\n",
       "    for (var i = 0; i < toolbar_spec.length; ++i) {\n",
       "        func = resolve_function(tool_api, toolbar_spec[i][\"setup\"][\"handler\"]);\n",
       "        toolbar_spec[i][\"setup\"][\"handler\"] = func;\n",
       "    }\n",
       "    return toolbar_spec\n",
       "}\n",
       "\n",
       "\n",
       "    var biolqm_jsapi = { action_upload_model: function() {\n",
       "        var cell = Jupyter.notebook.get_selected_cell();\n",
       "        cell.set_text('lqm = '+biolqm_jsapi.module_alias+'.upload()');\n",
       "        cell.focus_editor();\n",
       "    } };\n",
       "    colomoto_extension(Jupyter, \"colomoto-setup-5204324\", \"biolqm\", {\"name\": \"bioLQM\", \"sub-menu\": [{\"name\": \"Upload model\", \"snippet\": [\"lqm = biolqm.upload()\"]}, {\"name\": \"Load model\", \"snippet\": [\"lqm = biolqm.load(\\\"model.sbml\\\")\"]}, \"---\", {\"name\": \"Export to file\", \"sub-menu\": [{\"name\": \"SBML-qual v1.0 (.sbml)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.sbml\\\", \\\"sbml\\\")\"]}, \"Functions formats\", {\"name\": \"BoolNet (.bnet)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.bnet\\\", \\\"bnet\\\")\"]}, {\"name\": \"BooleanNet (.booleannet)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.booleannet\\\", \\\"booleannet\\\")\"]}, {\"name\": \"BoolSim (.boolsim)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.boolsim\\\", \\\"boolsim\\\")\"]}, {\"name\": \"Raw Boolean functions (.boolfunctions)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.boolfunctions\\\", \\\"boolfunctions\\\")\"]}, {\"name\": \"Truth table (.tt)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.tt\\\", \\\"tt\\\")\"]}, \"Petri net formats\", {\"name\": \"APNN (.apnn)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.apnn\\\", \\\"apnn\\\")\"]}, {\"name\": \"INA (.ina)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.ina\\\", \\\"ina\\\")\"]}, {\"name\": \"PNML (.pnml)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.pnml\\\", \\\"pnml\\\")\"]}, \"Dedicated formats\", {\"name\": \"GINML (.ginml)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.ginml\\\", \\\"ginml\\\")\"]}, {\"name\": \"GNA non-xml (.gna)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.gna\\\", \\\"gna\\\")\"]}, {\"name\": \"MaBoSS (.bnd)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.bnd\\\", \\\"bnd\\\")\"]}, {\"name\": \"Pint (.an)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.an\\\", \\\"an\\\")\"]}]}, {\"name\": \"Convert to tool\", \"sub-menu\": [{\"name\": \"GINsim\", \"snippet\": [\"lrg = biolqm.to_ginsim(lqm)\"]}, {\"name\": \"MaBoSS\", \"snippet\": [\"masim = biolqm.to_maboss(lqm)\"]}, {\"name\": \"Pint\", \"snippet\": [\"an = biolqm.to_pint(lqm)\"]}]}, \"---\", {\"name\": \"Compute fixpoints\", \"snippet\": [\"fps = biolqm.fixpoints(lqm)\"]}, {\"name\": \"Compute trap spaces\", \"snippet\": [\"traps = biolqm.trapspace(lqm)\"]}, \"---\", {\"name\": \"Model modifications\", \"sub-menu\": [{\"name\": \"Perturbation\", \"snippet\": [\"lqm_mod = biolqm.perturbation(lqm, \\\"node%0\\\")\"]}, {\"name\": \"Booleanization\", \"snippet\": [\"lqm_bool = biolqm.booleanize(lqm)\"]}, {\"name\": \"Reduction\", \"snippet\": [\"lqm_red = biolqm.reduce(lqm, \\\"fixed,output,duplicate\\\")\"]}, {\"name\": \"Reversal\", \"snippet\": [\"lqm_rev = biolqm.reverse(lqm)\"]}, {\"name\": \"Sanitize\", \"snippet\": [\"lqm_san = biolqm.sanitize(lqm)\"]}]}, \"---\", {\"name\": \"Documentation\", \"external-link\": \"http://ginsim.naldi.info/biolqm/site/doc/index.html\"}]}, [{\"name\": \"upload\", \"setup\": {\"icon\": \"fa-upload\", \"help\": \"Upload model\", \"handler\": \"action_upload_model\"}}], biolqm_jsapi);\n",
       "     }</script><style type=\"text/css\">\n",
       "    #biolqm_menu {\n",
       "        color: #00007f;\n",
       "        background-color: white;\n",
       "    }\n",
       "    \n",
       "    div#biolqm-toolbar:before {\n",
       "        content: \"bioLQM\";\n",
       "        float: left;\n",
       "        margin-left: 1em;\n",
       "        margin-right: 1em;\n",
       "        margin-top: 0.2em;\n",
       "    }\n",
       "    div#biolqm-toolbar,\n",
       "    div#biolqm-toolbar button {\n",
       "        color: #00007f;\n",
       "    }\n",
       "    </style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\" id=\"colomoto-setup-8256972\" class=\"to-be-removed\">\n",
       "        if (typeof Jupyter != 'undefined') {\n",
       "            \n",
       "    var ginsim_jsapi = { action_upload_model: function() {\n",
       "        var cell = Jupyter.notebook.get_selected_cell();\n",
       "        cell.set_text('lrg = '+ginsim_jsapi.module_alias+'.upload()');\n",
       "        cell.focus_editor();\n",
       "    } };\n",
       "    colomoto_extension(Jupyter, \"colomoto-setup-8256972\", \"ginsim\", {\"name\": \"GINsim\", \"sub-menu\": [{\"name\": \"Upload model\", \"snippet\": [\"lrg = ginsim.upload()\"]}, {\"name\": \"Load model\", \"snippet\": [\"lrg = ginsim.load(\\\"model.zginml\\\")\"]}, \"---\", {\"name\": \"Export to file\", \"sub-menu\": [{\"name\": \"MaBoSS (.bnd, .cfg)\", \"snippet\": [\"ginsim.service(\\\"maboss\\\").export(lrg, \\\"mymodel.bnd\\\")\"]}]}, {\"name\": \"Convert to tool\", \"sub-menu\": [{\"name\": \"bioLQM\", \"snippet\": [\"lqm = ginsim.to_biolqm(lrg)\"]}, {\"name\": \"MaBoSS\", \"snippet\": [\"masim = ginsim.to_maboss(lrg)\"]}, {\"name\": \"NuSMV\", \"snippet\": [\"smv = ginsim.to_nusmv(lrg)\"]}, {\"name\": \"Pint\", \"snippet\": [\"an = ginsim.to_pint(lrg)\"]}]}, \"---\", {\"name\": \" Display regulatory graph\", \"snippet\": [\"ginsim.show(lrg)\"]}]}, [{\"name\": \"upload\", \"setup\": {\"icon\": \"fa-upload\", \"help\": \"Upload model\", \"handler\": \"action_upload_model\"}}], ginsim_jsapi);\n",
       "     }</script><style type=\"text/css\">\n",
       "    #ginsim_menu {\n",
       "        color: blue;\n",
       "        background-color: white;\n",
       "    }\n",
       "    \n",
       "    div#ginsim-toolbar:before {\n",
       "        content: \"GINsim\";\n",
       "        float: left;\n",
       "        margin-left: 1em;\n",
       "        margin-right: 1em;\n",
       "        margin-top: 0.2em;\n",
       "    }\n",
       "    div#ginsim-toolbar,\n",
       "    div#ginsim-toolbar button {\n",
       "        color: blue;\n",
       "    }\n",
       "    </style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import logging\n",
    "\n",
    "# Add the project root to the Python path\n",
    "project_root = Path().absolute().parent\n",
    "sys.path.insert(0, str(project_root))\n",
    "\n",
    "# Import AAAIM functions\n",
    "from core import annotate_model, curate_model\n",
    "from core.model_info import (\n",
    "    find_species_with_chebi_annotations,\n",
    "    find_species_with_uniprot_annotations,\n",
    "    extract_model_info\n",
    ")\n",
    "from core.database_search import force_clear_chromadb, get_species_recommendations_rag\n",
    "from utils.evaluation import (\n",
    "    evaluate_single_model,\n",
    "    evaluate_models_in_folder,\n",
    "    print_evaluation_results,\n",
    "    compare_results,\n",
    "    process_saved_llm_responses,\n",
    "    analyze_bqbiol_qualifier_statistics,\n",
    "    filter_qualifiers_in_results\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotation distribution across bqbiol qualifiers and ontologies\n",
    "\n",
    "Count the number of species that contain each ontology term for each bqbiol qualifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing 1063 models for bqbiol qualifier statistics...\n",
      "Processing BIOMD0000000634.xml...\n",
      "Processing BIOMD0000000152.xml...\n",
      "Processing BIOMD0000000146.xml...\n",
      "Processing BIOMD0000000620.xml...\n",
      "Processing BIOMD0000000608.xml...\n",
      "Processing BIOMD0000000191.xml...\n",
      "Processing BIOMD0000000185.xml...\n",
      "Processing BIOMD0000000813.xml...\n",
      "Processing BIOMD0000000807.xml...\n",
      "Processing BIOMD0000000350.xml...\n",
      "Processing BIOMD0000000436.xml...\n",
      "Processing BIOMD0000000422.xml...\n",
      "Processing BIOMD0000000344.xml...\n",
      "Processing BIOMD0000001072.xml...\n",
      "Processing BIOMD0000000378.xml...\n",
      "Processing BIOMD0000000393.xml...\n",
      "Processing BIOMD0000000387.xml...\n",
      "Processing BIOMD0000000224.xml...\n",
      "Processing BIOMD0000000542.xml...\n",
      "Processing BIOMD0000000556.xml...\n",
      "Processing BIOMD0000000230.xml...\n",
      "Processing BIOMD0000000218.xml...\n",
      "Processing BIOMD0000000581.xml...\n",
      "Processing BIOMD0000000595.xml...\n",
      "Processing BIOMD0000000740.xml...\n",
      "Processing BIOMD0000000998.xml...\n",
      "Processing BIOMD0000000026.xml...\n",
      "Processing BIOMD0000000032.xml...\n",
      "Processing BIOMD0000000754.xml...\n",
      "Processing BIOMD0000000768.xml...\n",
      "Processing BIOMD0000000783.xml...\n",
      "Processing BIOMD0000000797.xml...\n",
      "Processing BIOMD0000000967.xml...\n",
      "Processing BIOMD0000000973.xml...\n",
      "Processing BIOMD0000000972.xml...\n",
      "Processing BIOMD0000000966.xml...\n",
      "Processing BIOMD0000000796.xml...\n",
      "Processing BIOMD0000000782.xml...\n",
      "Processing BIOMD0000000769.xml...\n",
      "Processing BIOMD0000000033.xml...\n",
      "Processing BIOMD0000000755.xml...\n",
      "Processing BIOMD0000000999.xml...\n",
      "Processing BIOMD0000000741.xml...\n",
      "Processing BIOMD0000000027.xml...\n",
      "Processing BIOMD0000000594.xml...\n",
      "Processing BIOMD0000000580.xml...\n",
      "Processing BIOMD0000000219.xml...\n",
      "Processing BIOMD0000000557.xml...\n",
      "Processing BIOMD0000000231.xml...\n",
      "Processing BIOMD0000000225.xml...\n",
      "Processing BIOMD0000000543.xml...\n",
      "Processing BIOMD0000000386.xml...\n",
      "Processing BIOMD0000000392.xml...\n",
      "Processing BIOMD0000000379.xml...\n",
      "Processing BIOMD0000000423.xml...\n",
      "Processing BIOMD0000000345.xml...\n",
      "Processing BIOMD0000000351.xml...\n",
      "Processing BIOMD0000000437.xml...\n",
      "Processing BIOMD0000000806.xml...\n",
      "Processing BIOMD0000000812.xml...\n",
      "Processing BIOMD0000000184.xml...\n",
      "Processing BIOMD0000000190.xml...\n",
      "Processing BIOMD0000000609.xml...\n",
      "Processing BIOMD0000000147.xml...\n",
      "Processing BIOMD0000000621.xml...\n",
      "Processing BIOMD0000000635.xml...\n",
      "Processing BIOMD0000000153.xml...\n",
      "Processing BIOMD0000000623.xml...\n",
      "Processing BIOMD0000000145.xml...\n",
      "Processing BIOMD0000000151.xml...\n",
      "Processing BIOMD0000000637.xml...\n",
      "Processing BIOMD0000000179.xml...\n",
      "Processing BIOMD0000000838.xml...\n",
      "Processing BIOMD0000000186.xml...\n",
      "Processing BIOMD0000000192.xml...\n",
      "Processing BIOMD0000000804.xml...\n",
      "Processing BIOMD0000000810.xml...\n",
      "Processing BIOMD0000001059.xml...\n",
      "Processing BIOMD0000000347.xml...\n",
      "Processing BIOMD0000000421.xml...\n",
      "Processing BIOMD0000000435.xml...\n",
      "Processing BIOMD0000000353.xml...\n",
      "Processing BIOMD0000001065.xml...\n",
      "Processing BIOMD0000000409.xml...\n",
      "Processing BIOMD0000000384.xml...\n",
      "Processing BIOMD0000000390.xml...\n",
      "Processing BIOMD0000000233.xml...\n",
      "Processing BIOMD0000000555.xml...\n",
      "Processing BIOMD0000000541.xml...\n",
      "Processing BIOMD0000000227.xml...\n",
      "Processing BIOMD0000000569.xml...\n",
      "Processing BIOMD0000000596.xml...\n",
      "Processing BIOMD0000000582.xml...\n",
      "Processing BIOMD0000000757.xml...\n",
      "Processing BIOMD0000000031.xml...\n",
      "Processing BIOMD0000000025.xml...\n",
      "Processing BIOMD0000000743.xml...\n",
      "Processing BIOMD0000000019.xml...\n",
      "Processing BIOMD0000000794.xml...\n",
      "Processing BIOMD0000000958.xml...\n",
      "Processing BIOMD0000000780.xml...\n",
      "Processing BIOMD0000000970.xml...\n",
      "Processing BIOMD0000000964.xml...\n",
      "Processing BIOMD0000000965.xml...\n",
      "Processing BIOMD0000000971.xml...\n",
      "Processing BIOMD0000000781.xml...\n",
      "Processing BIOMD0000000959.xml...\n",
      "Processing BIOMD0000000795.xml...\n",
      "Processing BIOMD0000000018.xml...\n",
      "Processing BIOMD0000000024.xml...\n",
      "Processing BIOMD0000000742.xml...\n",
      "Processing BIOMD0000000756.xml...\n",
      "Processing BIOMD0000000030.xml...\n",
      "Processing BIOMD0000000583.xml...\n",
      "Processing BIOMD0000000597.xml...\n",
      "Processing BIOMD0000000568.xml...\n",
      "Processing BIOMD0000000540.xml...\n",
      "Processing BIOMD0000000226.xml...\n",
      "Processing BIOMD0000000232.xml...\n",
      "Processing BIOMD0000000554.xml...\n",
      "Processing BIOMD0000000391.xml...\n",
      "Processing BIOMD0000000385.xml...\n",
      "Processing BIOMD0000000408.xml...\n",
      "Processing BIOMD0000001064.xml...\n",
      "Processing BIOMD0000000434.xml...\n",
      "Processing BIOMD0000000352.xml...\n",
      "Processing BIOMD0000000346.xml...\n",
      "Processing BIOMD0000001058.xml...\n",
      "Processing BIOMD0000000420.xml...\n",
      "Processing BIOMD0000000811.xml...\n",
      "Processing BIOMD0000000805.xml...\n",
      "Processing BIOMD0000000193.xml...\n",
      "Processing BIOMD0000000839.xml...\n",
      "Processing BIOMD0000000187.xml...\n",
      "Processing BIOMD0000000178.xml...\n",
      "Processing BIOMD0000000150.xml...\n",
      "Processing BIOMD0000000636.xml...\n",
      "Processing BIOMD0000000622.xml...\n",
      "Processing BIOMD0000000144.xml...\n",
      "Processing BIOMD0000000168.xml...\n",
      "Processing BIOMD0000000140.xml...\n",
      "Processing BIOMD0000000626.xml...\n",
      "Processing BIOMD0000000632.xml...\n",
      "Processing BIOMD0000000154.xml...\n",
      "Processing BIOMD0000000801.xml...\n",
      "Processing BIOMD0000000815.xml...\n",
      "Processing BIOMD0000000183.xml...\n",
      "Processing BIOMD0000000829.xml...\n",
      "Processing BIOMD0000000197.xml...\n",
      "Processing BIOMD0000000418.xml...\n",
      "Processing BIOMD0000001060.xml...\n",
      "Processing BIOMD0000000424.xml...\n",
      "Processing BIOMD0000000342.xml...\n",
      "Processing BIOMD0000001048.xml...\n",
      "Processing BIOMD0000000356.xml...\n",
      "Processing BIOMD0000000430.xml...\n",
      "Processing BIOMD0000000381.xml...\n",
      "Processing BIOMD0000000395.xml...\n",
      "Processing BIOMD0000000578.xml...\n",
      "Processing BIOMD0000000550.xml...\n",
      "Processing BIOMD0000000236.xml...\n",
      "Processing BIOMD0000000222.xml...\n",
      "Processing BIOMD0000000544.xml...\n",
      "Processing BIOMD0000000593.xml...\n",
      "Processing BIOMD0000000587.xml...\n",
      "Processing BIOMD0000000008.xml...\n",
      "Processing BIOMD0000000034.xml...\n",
      "Processing BIOMD0000000752.xml...\n",
      "Processing BIOMD0000000746.xml...\n",
      "Processing BIOMD0000000020.xml...\n",
      "Processing BIOMD0000000975.xml...\n",
      "Processing BIOMD0000000961.xml...\n",
      "Processing BIOMD0000000949.xml...\n",
      "Processing BIOMD0000000791.xml...\n",
      "Processing BIOMD0000000785.xml...\n",
      "Processing BIOMD0000000784.xml...\n",
      "Processing BIOMD0000000790.xml...\n",
      "Processing BIOMD0000000948.xml...\n",
      "Processing BIOMD0000000960.xml...\n",
      "Processing BIOMD0000000974.xml...\n",
      "Processing BIOMD0000000747.xml...\n",
      "Processing BIOMD0000000021.xml...\n",
      "Processing BIOMD0000000035.xml...\n",
      "Processing BIOMD0000000753.xml...\n",
      "Processing BIOMD0000000009.xml...\n",
      "Processing BIOMD0000000586.xml...\n",
      "Processing BIOMD0000000592.xml...\n",
      "Processing BIOMD0000000223.xml...\n",
      "Processing BIOMD0000000545.xml...\n",
      "Processing BIOMD0000000551.xml...\n",
      "Processing BIOMD0000000237.xml...\n",
      "Processing BIOMD0000000579.xml...\n",
      "Processing BIOMD0000000394.xml...\n",
      "Processing BIOMD0000000380.xml...\n",
      "Processing BIOMD0000000357.xml...\n",
      "Processing BIOMD0000000431.xml...\n",
      "Processing BIOMD0000000425.xml...\n",
      "Processing BIOMD0000000343.xml...\n",
      "Processing BIOMD0000000419.xml...\n",
      "Processing BIOMD0000001061.xml...\n",
      "Processing BIOMD0000000828.xml...\n",
      "Processing BIOMD0000000196.xml...\n",
      "Processing BIOMD0000000182.xml...\n",
      "Processing BIOMD0000000814.xml...\n",
      "Processing BIOMD0000000800.xml...\n",
      "Processing BIOMD0000000633.xml...\n",
      "Processing BIOMD0000000155.xml...\n",
      "Processing BIOMD0000000141.xml...\n",
      "Processing BIOMD0000000627.xml...\n",
      "Processing BIOMD0000000169.xml...\n",
      "Processing BIOMD0000000619.xml...\n",
      "Processing BIOMD0000000157.xml...\n",
      "Processing BIOMD0000000631.xml...\n",
      "Processing BIOMD0000000625.xml...\n",
      "Processing BIOMD0000000143.xml...\n",
      "Processing BIOMD0000000816.xml...\n",
      "Processing BIOMD0000000802.xml...\n",
      "Processing BIOMD0000000194.xml...\n",
      "Processing BIOMD0000000180.xml...\n",
      "Processing BIOMD0000001077.xml...\n",
      "Processing BIOMD0000000369.xml...\n",
      "Processing BIOMD0000001063.xml...\n",
      "Processing BIOMD0000000433.xml...\n",
      "Processing BIOMD0000000355.xml...\n",
      "Processing BIOMD0000000341.xml...\n",
      "Processing BIOMD0000000427.xml...\n",
      "Processing BIOMD0000000396.xml...\n",
      "Processing BIOMD0000000382.xml...\n",
      "Processing BIOMD0000000209.xml...\n",
      "Processing BIOMD0000000547.xml...\n",
      "Processing BIOMD0000000221.xml...\n",
      "Processing BIOMD0000000235.xml...\n",
      "Processing BIOMD0000000553.xml...\n",
      "Processing BIOMD0000000584.xml...\n",
      "Processing BIOMD0000000590.xml...\n",
      "Processing BIOMD0000000779.xml...\n",
      "Processing BIOMD0000000023.xml...\n",
      "Processing BIOMD0000000745.xml...\n",
      "Processing BIOMD0000000751.xml...\n",
      "Processing BIOMD0000000989.xml...\n",
      "Processing BIOMD0000000037.xml...\n",
      "Processing BIOMD0000000962.xml...\n",
      "Processing BIOMD0000000976.xml...\n",
      "Processing BIOMD0000000786.xml...\n",
      "Processing BIOMD0000000792.xml...\n",
      "Processing BIOMD0000000793.xml...\n",
      "Processing BIOMD0000000787.xml...\n",
      "Processing BIOMD0000000977.xml...\n",
      "Processing BIOMD0000000963.xml...\n",
      "Processing BIOMD0000000988.xml...\n",
      "Processing BIOMD0000000750.xml...\n",
      "Processing BIOMD0000000036.xml...\n",
      "Processing BIOMD0000000022.xml...\n",
      "Processing BIOMD0000000744.xml...\n",
      "Processing BIOMD0000000778.xml...\n",
      "Processing BIOMD0000000591.xml...\n",
      "Processing BIOMD0000000585.xml...\n",
      "Processing BIOMD0000000234.xml...\n",
      "Processing BIOMD0000000552.xml...\n",
      "Processing BIOMD0000000546.xml...\n",
      "Processing BIOMD0000000220.xml...\n",
      "Processing BIOMD0000000208.xml...\n",
      "Processing BIOMD0000000383.xml...\n",
      "Processing BIOMD0000000397.xml...\n",
      "Processing BIOMD0000000340.xml...\n",
      "Processing BIOMD0000000426.xml...\n",
      "Processing BIOMD0000000432.xml...\n",
      "Processing BIOMD0000000354.xml...\n",
      "Processing BIOMD0000001062.xml...\n",
      "Processing BIOMD0000000368.xml...\n",
      "Processing BIOMD0000000181.xml...\n",
      "Processing BIOMD0000000195.xml...\n",
      "Processing BIOMD0000000803.xml...\n",
      "Processing BIOMD0000000817.xml...\n",
      "Processing BIOMD0000000624.xml...\n",
      "Processing BIOMD0000000142.xml...\n",
      "Processing BIOMD0000000156.xml...\n",
      "Processing BIOMD0000000630.xml...\n",
      "Processing BIOMD0000000618.xml...\n",
      "Processing BIOMD0000000131.xml...\n",
      "Processing BIOMD0000000657.xml...\n",
      "Processing BIOMD0000000643.xml...\n",
      "Processing BIOMD0000000125.xml...\n",
      "Processing BIOMD0000000119.xml...\n",
      "Processing BIOMD0000000680.xml...\n",
      "Processing BIOMD0000000858.xml...\n",
      "Processing BIOMD0000000870.xml...\n",
      "Processing BIOMD0000000864.xml...\n",
      "Processing BIOMD0000000455.xml...\n",
      "Processing BIOMD0000000333.xml...\n",
      "Processing BIOMD0000001039.xml...\n",
      "Processing BIOMD0000000327.xml...\n",
      "Processing BIOMD0000000441.xml...\n",
      "Processing BIOMD0000000469.xml...\n",
      "Processing BIOMD0000001011.xml...\n",
      "Processing BIOMD0000001005.xml...\n",
      "Processing BIOMD0000000496.xml...\n",
      "Processing BIOMD0000000482.xml...\n",
      "Processing BIOMD0000000521.xml...\n",
      "Processing BIOMD0000000247.xml...\n",
      "Processing BIOMD0000000253.xml...\n",
      "Processing BIOMD0000000535.xml...\n",
      "Processing BIOMD0000000509.xml...\n",
      "Processing BIOMD0000000284.xml...\n",
      "Processing BIOMD0000000290.xml...\n",
      "Processing BIOMD0000000045.xml...\n",
      "Processing BIOMD0000000723.xml...\n",
      "Processing BIOMD0000000737.xml...\n",
      "Processing BIOMD0000000051.xml...\n",
      "Processing BIOMD0000000079.xml...\n",
      "Processing BIOMD0000000086.xml...\n",
      "Processing BIOMD0000000938.xml...\n",
      "Processing BIOMD0000000092.xml...\n",
      "Processing BIOMD0000000904.xml...\n",
      "Processing BIOMD0000000910.xml...\n",
      "Processing BIOMD0000000911.xml...\n",
      "Processing BIOMD0000000905.xml...\n",
      "Processing BIOMD0000000093.xml...\n",
      "Processing BIOMD0000000087.xml...\n",
      "Processing BIOMD0000000939.xml...\n",
      "Processing BIOMD0000000078.xml...\n",
      "Processing BIOMD0000000736.xml...\n",
      "Processing BIOMD0000000050.xml...\n",
      "Processing BIOMD0000000044.xml...\n",
      "Processing BIOMD0000000722.xml...\n",
      "Processing BIOMD0000000291.xml...\n",
      "Processing BIOMD0000000285.xml...\n",
      "Processing BIOMD0000000508.xml...\n",
      "Processing BIOMD0000000252.xml...\n",
      "Processing BIOMD0000000534.xml...\n",
      "Processing BIOMD0000000520.xml...\n",
      "Processing BIOMD0000000246.xml...\n",
      "Processing BIOMD0000000483.xml...\n",
      "Processing BIOMD0000000497.xml...\n",
      "Processing BIOMD0000001004.xml...\n",
      "Processing BIOMD0000000468.xml...\n",
      "Processing BIOMD0000001010.xml...\n",
      "Processing BIOMD0000000326.xml...\n",
      "Processing BIOMD0000001038.xml...\n",
      "Processing BIOMD0000000440.xml...\n",
      "Processing BIOMD0000000454.xml...\n",
      "Processing BIOMD0000000332.xml...\n",
      "Processing BIOMD0000000865.xml...\n",
      "Processing BIOMD0000000871.xml...\n",
      "Processing BIOMD0000000859.xml...\n",
      "Processing BIOMD0000000681.xml...\n",
      "Processing BIOMD0000000695.xml...\n",
      "Processing BIOMD0000000118.xml...\n",
      "Processing BIOMD0000000642.xml...\n",
      "Processing BIOMD0000000124.xml...\n",
      "Processing BIOMD0000000130.xml...\n",
      "Processing BIOMD0000000656.xml...\n",
      "Processing BIOMD0000000126.xml...\n",
      "Processing BIOMD0000000898.xml...\n",
      "Processing BIOMD0000000640.xml...\n",
      "Processing BIOMD0000000654.xml...\n",
      "Processing BIOMD0000000132.xml...\n",
      "Processing BIOMD0000000668.xml...\n",
      "Processing BIOMD0000000683.xml...\n",
      "Processing BIOMD0000000697.xml...\n",
      "Processing BIOMD0000000867.xml...\n",
      "Processing BIOMD0000000873.xml...\n",
      "Processing BIOMD0000000442.xml...\n",
      "Processing BIOMD0000000324.xml...\n",
      "Processing BIOMD0000000330.xml...\n",
      "Processing BIOMD0000000456.xml...\n",
      "Processing BIOMD0000001006.xml...\n",
      "Processing BIOMD0000000318.xml...\n",
      "Processing BIOMD0000001012.xml...\n",
      "Processing BIOMD0000000481.xml...\n",
      "Processing BIOMD0000000495.xml...\n",
      "Processing BIOMD0000000536.xml...\n",
      "Processing BIOMD0000000250.xml...\n",
      "Processing BIOMD0000000244.xml...\n",
      "Processing BIOMD0000000522.xml...\n",
      "Processing BIOMD0000000278.xml...\n",
      "Processing BIOMD0000000293.xml...\n",
      "Processing BIOMD0000000287.xml...\n",
      "Processing BIOMD0000000052.xml...\n",
      "Processing BIOMD0000000734.xml...\n",
      "Processing BIOMD0000000720.xml...\n",
      "Processing BIOMD0000000046.xml...\n",
      "Processing BIOMD0000000708.xml...\n",
      "Processing BIOMD0000000091.xml...\n",
      "Processing BIOMD0000000085.xml...\n",
      "Processing BIOMD0000000913.xml...\n",
      "Processing BIOMD0000000907.xml...\n",
      "Processing BIOMD0000000906.xml...\n",
      "Processing BIOMD0000000912.xml...\n",
      "Processing BIOMD0000000084.xml...\n",
      "Processing BIOMD0000000090.xml...\n",
      "Processing BIOMD0000000709.xml...\n",
      "Processing BIOMD0000000721.xml...\n",
      "Processing BIOMD0000000047.xml...\n",
      "Processing BIOMD0000000053.xml...\n",
      "Processing BIOMD0000000735.xml...\n",
      "Processing BIOMD0000000286.xml...\n",
      "Processing BIOMD0000000292.xml...\n",
      "Processing BIOMD0000000279.xml...\n",
      "Processing BIOMD0000000245.xml...\n",
      "Processing BIOMD0000000523.xml...\n",
      "Processing BIOMD0000000537.xml...\n",
      "Processing BIOMD0000000251.xml...\n",
      "Processing BIOMD0000000494.xml...\n",
      "Processing BIOMD0000000480.xml...\n",
      "Processing BIOMD0000001013.xml...\n",
      "Processing BIOMD0000000319.xml...\n",
      "Processing BIOMD0000001007.xml...\n",
      "Processing BIOMD0000000331.xml...\n",
      "Processing BIOMD0000000457.xml...\n",
      "Processing BIOMD0000000443.xml...\n",
      "Processing BIOMD0000000325.xml...\n",
      "Processing BIOMD0000000872.xml...\n",
      "Processing BIOMD0000000866.xml...\n",
      "Processing BIOMD0000000696.xml...\n",
      "Processing BIOMD0000000682.xml...\n",
      "Processing BIOMD0000000669.xml...\n",
      "Processing BIOMD0000000655.xml...\n",
      "Processing BIOMD0000000133.xml...\n",
      "Processing BIOMD0000000127.xml...\n",
      "Processing BIOMD0000000641.xml...\n",
      "Processing BIOMD0000000899.xml...\n",
      "Processing BIOMD0000000679.xml...\n",
      "Processing BIOMD0000000645.xml...\n",
      "Processing BIOMD0000000123.xml...\n",
      "Processing BIOMD0000000137.xml...\n",
      "Processing BIOMD0000000889.xml...\n",
      "Processing BIOMD0000000651.xml...\n",
      "Processing BIOMD0000000862.xml...\n",
      "Processing BIOMD0000000876.xml...\n",
      "Processing BIOMD0000000686.xml...\n",
      "Processing BIOMD0000000692.xml...\n",
      "Processing BIOMD0000001003.xml...\n",
      "Processing BIOMD0000001017.xml...\n",
      "Processing BIOMD0000000309.xml...\n",
      "Processing BIOMD0000000321.xml...\n",
      "Processing BIOMD0000000447.xml...\n",
      "Processing BIOMD0000000453.xml...\n",
      "Processing BIOMD0000000335.xml...\n",
      "Processing BIOMD0000000484.xml...\n",
      "Processing BIOMD0000000490.xml...\n",
      "Processing BIOMD0000000269.xml...\n",
      "Processing BIOMD0000000255.xml...\n",
      "Processing BIOMD0000000533.xml...\n",
      "Processing BIOMD0000000527.xml...\n",
      "Processing BIOMD0000000241.xml...\n",
      "Processing BIOMD0000000296.xml...\n",
      "Processing BIOMD0000000282.xml...\n",
      "Processing BIOMD0000000719.xml...\n",
      "Processing BIOMD0000000731.xml...\n",
      "Processing BIOMD0000000057.xml...\n",
      "Processing BIOMD0000000043.xml...\n",
      "Processing BIOMD0000000725.xml...\n",
      "Processing BIOMD0000000916.xml...\n",
      "Processing BIOMD0000000902.xml...\n",
      "Processing BIOMD0000000094.xml...\n",
      "Processing BIOMD0000000080.xml...\n",
      "Processing BIOMD0000000081.xml...\n",
      "Processing BIOMD0000000095.xml...\n",
      "Processing BIOMD0000000903.xml...\n",
      "Processing BIOMD0000000917.xml...\n",
      "Processing BIOMD0000000042.xml...\n",
      "Processing BIOMD0000000724.xml...\n",
      "Processing BIOMD0000000730.xml...\n",
      "Processing BIOMD0000000056.xml...\n",
      "Processing BIOMD0000000718.xml...\n",
      "Processing BIOMD0000000283.xml...\n",
      "Processing BIOMD0000000297.xml...\n",
      "Processing BIOMD0000000526.xml...\n",
      "Processing BIOMD0000000240.xml...\n",
      "Processing BIOMD0000000254.xml...\n",
      "Processing BIOMD0000000532.xml...\n",
      "Processing BIOMD0000000268.xml...\n",
      "Processing BIOMD0000000491.xml...\n",
      "Processing BIOMD0000000485.xml...\n",
      "Processing BIOMD0000000452.xml...\n",
      "Processing BIOMD0000000334.xml...\n",
      "Processing BIOMD0000000320.xml...\n",
      "Processing BIOMD0000000446.xml...\n",
      "Processing BIOMD0000000308.xml...\n",
      "Processing BIOMD0000001016.xml...\n",
      "Processing BIOMD0000001002.xml...\n",
      "Processing BIOMD0000000693.xml...\n",
      "Processing BIOMD0000000687.xml...\n",
      "Processing BIOMD0000000877.xml...\n",
      "Processing BIOMD0000000863.xml...\n",
      "Processing BIOMD0000000136.xml...\n",
      "Processing BIOMD0000000650.xml...\n",
      "Processing BIOMD0000000888.xml...\n",
      "Processing BIOMD0000000644.xml...\n",
      "Processing BIOMD0000000122.xml...\n",
      "Processing BIOMD0000000678.xml...\n",
      "Processing BIOMD0000000108.xml...\n",
      "Processing BIOMD0000000652.xml...\n",
      "Processing BIOMD0000000134.xml...\n",
      "Processing BIOMD0000000120.xml...\n",
      "Processing BIOMD0000000646.xml...\n",
      "Processing BIOMD0000000875.xml...\n",
      "Processing BIOMD0000000861.xml...\n",
      "Processing BIOMD0000000691.xml...\n",
      "Processing BIOMD0000000849.xml...\n",
      "Processing BIOMD0000000685.xml...\n",
      "Processing BIOMD0000001014.xml...\n",
      "Processing BIOMD0000000478.xml...\n",
      "Processing BIOMD0000001000.xml...\n",
      "Processing BIOMD0000001028.xml...\n",
      "Processing BIOMD0000000336.xml...\n",
      "Processing BIOMD0000000450.xml...\n",
      "Processing BIOMD0000000444.xml...\n",
      "Processing BIOMD0000000322.xml...\n",
      "Processing BIOMD0000000493.xml...\n",
      "Processing BIOMD0000000487.xml...\n",
      "Processing BIOMD0000000518.xml...\n",
      "Processing BIOMD0000000242.xml...\n",
      "Processing BIOMD0000000524.xml...\n",
      "Processing BIOMD0000000530.xml...\n",
      "Processing BIOMD0000000256.xml...\n",
      "Processing BIOMD0000000281.xml...\n",
      "Processing BIOMD0000000295.xml...\n",
      "Processing BIOMD0000000068.xml...\n",
      "Processing BIOMD0000000726.xml...\n",
      "Processing BIOMD0000000040.xml...\n",
      "Processing BIOMD0000000054.xml...\n",
      "Processing BIOMD0000000732.xml...\n",
      "Processing BIOMD0000000901.xml...\n",
      "Processing BIOMD0000000915.xml...\n",
      "Processing BIOMD0000000083.xml...\n",
      "Processing BIOMD0000000097.xml...\n",
      "Processing BIOMD0000000929.xml...\n",
      "Processing BIOMD0000000096.xml...\n",
      "Processing BIOMD0000000928.xml...\n",
      "Processing BIOMD0000000082.xml...\n",
      "Processing BIOMD0000000914.xml...\n",
      "Processing BIOMD0000000900.xml...\n",
      "Processing BIOMD0000000055.xml...\n",
      "Processing BIOMD0000000733.xml...\n",
      "Processing BIOMD0000000727.xml...\n",
      "Processing BIOMD0000000041.xml...\n",
      "Processing BIOMD0000000069.xml...\n",
      "Processing BIOMD0000000294.xml...\n",
      "Processing BIOMD0000000280.xml...\n",
      "Processing BIOMD0000000531.xml...\n",
      "Processing BIOMD0000000257.xml...\n",
      "Processing BIOMD0000000243.xml...\n",
      "Processing BIOMD0000000525.xml...\n",
      "Processing BIOMD0000000519.xml...\n",
      "Processing BIOMD0000000486.xml...\n",
      "Processing BIOMD0000000492.xml...\n",
      "Processing BIOMD0000000445.xml...\n",
      "Processing BIOMD0000000323.xml...\n",
      "Processing BIOMD0000000337.xml...\n",
      "Processing BIOMD0000001029.xml...\n",
      "Processing BIOMD0000000451.xml...\n",
      "Processing BIOMD0000000479.xml...\n",
      "Processing BIOMD0000001001.xml...\n",
      "Processing BIOMD0000001015.xml...\n",
      "Processing BIOMD0000000684.xml...\n",
      "Processing BIOMD0000000848.xml...\n",
      "Processing BIOMD0000000690.xml...\n",
      "Processing BIOMD0000000860.xml...\n",
      "Processing BIOMD0000000874.xml...\n",
      "Processing BIOMD0000000121.xml...\n",
      "Processing BIOMD0000000647.xml...\n",
      "Processing BIOMD0000000653.xml...\n",
      "Processing BIOMD0000000135.xml...\n",
      "Processing BIOMD0000000109.xml...\n",
      "Processing BIOMD0000000110.xml...\n",
      "Processing BIOMD0000000676.xml...\n",
      "Processing BIOMD0000000662.xml...\n",
      "Processing BIOMD0000000104.xml...\n",
      "Processing BIOMD0000000892.xml...\n",
      "Processing BIOMD0000000886.xml...\n",
      "Processing BIOMD0000000138.xml...\n",
      "Processing BIOMD0000000879.xml...\n",
      "Processing BIOMD0000000689.xml...\n",
      "Processing BIOMD0000000851.xml...\n",
      "Processing BIOMD0000000845.xml...\n",
      "Processing BIOMD0000000474.xml...\n",
      "Processing BIOMD0000000312.xml...\n",
      "Processing BIOMD0000001018.xml...\n",
      "Processing BIOMD0000000306.xml...\n",
      "Processing BIOMD0000000460.xml...\n",
      "Processing BIOMD0000000448.xml...\n",
      "Processing BIOMD0000001030.xml...\n",
      "Processing BIOMD0000001024.xml...\n",
      "Processing BIOMD0000000500.xml...\n",
      "Processing BIOMD0000000266.xml...\n",
      "Processing BIOMD0000000272.xml...\n",
      "Processing BIOMD0000000514.xml...\n",
      "Processing BIOMD0000000528.xml...\n",
      "Processing BIOMD0000000299.xml...\n",
      "Processing BIOMD0000000064.xml...\n",
      "Processing BIOMD0000000702.xml...\n",
      "Processing BIOMD0000000716.xml...\n",
      "Processing BIOMD0000000070.xml...\n",
      "Processing BIOMD0000000058.xml...\n",
      "Processing BIOMD0000000919.xml...\n",
      "Processing BIOMD0000000925.xml...\n",
      "Processing BIOMD0000000931.xml...\n",
      "Processing BIOMD0000000930.xml...\n",
      "Processing BIOMD0000000924.xml...\n",
      "Processing BIOMD0000000918.xml...\n",
      "Processing BIOMD0000000059.xml...\n",
      "Processing BIOMD0000000717.xml...\n",
      "Processing BIOMD0000000071.xml...\n",
      "Processing BIOMD0000000065.xml...\n",
      "Processing BIOMD0000000703.xml...\n",
      "Processing BIOMD0000000298.xml...\n",
      "Processing BIOMD0000000529.xml...\n",
      "Processing BIOMD0000000273.xml...\n",
      "Processing BIOMD0000000515.xml...\n",
      "Processing BIOMD0000000501.xml...\n",
      "Processing BIOMD0000000267.xml...\n",
      "Processing BIOMD0000001025.xml...\n",
      "Processing BIOMD0000000449.xml...\n",
      "Processing BIOMD0000001031.xml...\n",
      "Processing BIOMD0000000307.xml...\n",
      "Processing BIOMD0000001019.xml...\n",
      "Processing BIOMD0000000461.xml...\n",
      "Processing BIOMD0000000475.xml...\n",
      "Processing BIOMD0000000313.xml...\n",
      "Processing BIOMD0000000844.xml...\n",
      "Processing BIOMD0000000850.xml...\n",
      "Processing BIOMD0000000688.xml...\n",
      "Processing BIOMD0000000878.xml...\n",
      "Processing BIOMD0000000887.xml...\n",
      "Processing BIOMD0000000139.xml...\n",
      "Processing BIOMD0000000893.xml...\n",
      "Processing BIOMD0000000663.xml...\n",
      "Processing BIOMD0000000105.xml...\n",
      "Processing BIOMD0000000111.xml...\n",
      "Processing BIOMD0000000677.xml...\n",
      "Processing BIOMD0000000107.xml...\n",
      "Processing BIOMD0000000661.xml...\n",
      "Processing BIOMD0000000675.xml...\n",
      "Processing BIOMD0000000113.xml...\n",
      "Processing BIOMD0000000885.xml...\n",
      "Processing BIOMD0000000891.xml...\n",
      "Processing BIOMD0000000846.xml...\n",
      "Processing BIOMD0000000852.xml...\n",
      "Processing BIOMD0000000463.xml...\n",
      "Processing BIOMD0000000305.xml...\n",
      "Processing BIOMD0000000311.xml...\n",
      "Processing BIOMD0000000477.xml...\n",
      "Processing BIOMD0000001027.xml...\n",
      "Processing BIOMD0000000339.xml...\n",
      "Processing BIOMD0000001033.xml...\n",
      "Processing BIOMD0000000488.xml...\n",
      "Processing BIOMD0000000517.xml...\n",
      "Processing BIOMD0000000271.xml...\n",
      "Processing BIOMD0000000265.xml...\n",
      "Processing BIOMD0000000503.xml...\n",
      "Processing BIOMD0000000259.xml...\n",
      "Processing BIOMD0000000073.xml...\n",
      "Processing BIOMD0000000715.xml...\n",
      "Processing BIOMD0000000701.xml...\n",
      "Processing BIOMD0000000067.xml...\n",
      "Processing BIOMD0000000729.xml...\n",
      "Processing BIOMD0000000932.xml...\n",
      "Processing BIOMD0000000926.xml...\n",
      "Processing BIOMD0000000098.xml...\n",
      "Processing BIOMD0000000927.xml...\n",
      "Processing BIOMD0000000099.xml...\n",
      "Processing BIOMD0000000933.xml...\n",
      "Processing BIOMD0000000728.xml...\n",
      "Processing BIOMD0000000700.xml...\n",
      "Processing BIOMD0000000066.xml...\n",
      "Processing BIOMD0000000072.xml...\n",
      "Processing BIOMD0000000714.xml...\n",
      "Processing BIOMD0000000258.xml...\n",
      "Processing BIOMD0000000264.xml...\n",
      "Processing BIOMD0000000502.xml...\n",
      "Processing BIOMD0000000516.xml...\n",
      "Processing BIOMD0000000270.xml...\n",
      "Processing BIOMD0000000489.xml...\n",
      "Processing BIOMD0000001032.xml...\n",
      "Processing BIOMD0000000338.xml...\n",
      "Processing BIOMD0000001026.xml...\n",
      "Processing BIOMD0000000310.xml...\n",
      "Processing BIOMD0000000476.xml...\n",
      "Processing BIOMD0000000462.xml...\n",
      "Processing BIOMD0000000304.xml...\n",
      "Processing BIOMD0000000853.xml...\n",
      "Processing BIOMD0000000847.xml...\n",
      "Processing BIOMD0000000648.xml...\n",
      "Processing BIOMD0000000890.xml...\n",
      "Processing BIOMD0000000884.xml...\n",
      "Processing BIOMD0000000674.xml...\n",
      "Processing BIOMD0000000112.xml...\n",
      "Processing BIOMD0000000106.xml...\n",
      "Processing BIOMD0000000660.xml...\n",
      "Processing BIOMD0000000880.xml...\n",
      "Processing BIOMD0000000658.xml...\n",
      "Processing BIOMD0000000894.xml...\n",
      "Processing BIOMD0000000664.xml...\n",
      "Processing BIOMD0000000102.xml...\n",
      "Processing BIOMD0000000116.xml...\n",
      "Processing BIOMD0000000670.xml...\n",
      "Processing BIOMD0000000843.xml...\n",
      "Processing BIOMD0000000857.xml...\n",
      "Processing BIOMD0000001022.xml...\n",
      "Processing BIOMD0000001036.xml...\n",
      "Processing BIOMD0000000328.xml...\n",
      "Processing BIOMD0000000300.xml...\n",
      "Processing BIOMD0000000466.xml...\n",
      "Processing BIOMD0000000472.xml...\n",
      "Processing BIOMD0000000314.xml...\n",
      "Processing BIOMD0000000499.xml...\n",
      "Processing BIOMD0000000248.xml...\n",
      "Processing BIOMD0000000274.xml...\n",
      "Processing BIOMD0000000512.xml...\n",
      "Processing BIOMD0000000506.xml...\n",
      "Processing BIOMD0000000260.xml...\n",
      "Processing BIOMD0000000738.xml...\n",
      "Processing BIOMD0000000710.xml...\n",
      "Processing BIOMD0000000076.xml...\n",
      "Processing BIOMD0000000062.xml...\n",
      "Processing BIOMD0000000704.xml...\n",
      "Processing BIOMD0000000937.xml...\n",
      "Processing BIOMD0000000089.xml...\n",
      "Processing BIOMD0000000923.xml...\n",
      "Processing BIOMD0000000922.xml...\n",
      "Processing BIOMD0000000936.xml...\n",
      "Processing BIOMD0000000088.xml...\n",
      "Processing BIOMD0000000063.xml...\n",
      "Processing BIOMD0000000705.xml...\n",
      "Processing BIOMD0000000711.xml...\n",
      "Processing BIOMD0000000077.xml...\n",
      "Processing BIOMD0000000739.xml...\n",
      "Processing BIOMD0000000507.xml...\n",
      "Processing BIOMD0000000261.xml...\n",
      "Processing BIOMD0000000275.xml...\n",
      "Processing BIOMD0000000513.xml...\n",
      "Processing BIOMD0000000249.xml...\n",
      "Processing BIOMD0000000498.xml...\n",
      "Processing BIOMD0000000473.xml...\n",
      "Processing BIOMD0000000315.xml...\n",
      "Processing BIOMD0000000301.xml...\n",
      "Processing BIOMD0000000467.xml...\n",
      "Processing BIOMD0000000329.xml...\n",
      "Processing BIOMD0000001037.xml...\n",
      "Processing BIOMD0000001023.xml...\n",
      "Processing BIOMD0000000856.xml...\n",
      "Processing BIOMD0000000842.xml...\n",
      "Processing BIOMD0000000117.xml...\n",
      "Processing BIOMD0000000671.xml...\n",
      "Processing BIOMD0000000665.xml...\n",
      "Processing BIOMD0000000103.xml...\n",
      "Processing BIOMD0000000895.xml...\n",
      "Processing BIOMD0000000659.xml...\n",
      "Processing BIOMD0000000881.xml...\n",
      "Processing BIOMD0000000897.xml...\n",
      "Processing BIOMD0000000129.xml...\n",
      "Processing BIOMD0000000883.xml...\n",
      "Processing BIOMD0000000673.xml...\n",
      "Processing BIOMD0000000115.xml...\n",
      "Processing BIOMD0000000101.xml...\n",
      "Processing BIOMD0000000667.xml...\n",
      "Processing BIOMD0000000854.xml...\n",
      "Processing BIOMD0000000698.xml...\n",
      "Processing BIOMD0000000840.xml...\n",
      "Processing BIOMD0000000868.xml...\n",
      "Processing BIOMD0000001035.xml...\n",
      "Processing BIOMD0000000459.xml...\n",
      "Processing BIOMD0000001021.xml...\n",
      "Processing BIOMD0000001009.xml...\n",
      "Processing BIOMD0000000317.xml...\n",
      "Processing BIOMD0000000471.xml...\n",
      "Processing BIOMD0000000465.xml...\n",
      "Processing BIOMD0000000303.xml...\n",
      "Processing BIOMD0000000539.xml...\n",
      "Processing BIOMD0000000263.xml...\n",
      "Processing BIOMD0000000505.xml...\n",
      "Processing BIOMD0000000511.xml...\n",
      "Processing BIOMD0000000277.xml...\n",
      "Processing BIOMD0000000288.xml...\n",
      "Processing BIOMD0000000049.xml...\n",
      "Processing BIOMD0000000707.xml...\n",
      "Processing BIOMD0000000061.xml...\n",
      "Processing BIOMD0000000075.xml...\n",
      "Processing BIOMD0000000713.xml...\n",
      "Processing BIOMD0000000920.xml...\n",
      "Processing BIOMD0000000934.xml...\n",
      "Processing BIOMD0000000908.xml...\n",
      "Processing BIOMD0000000909.xml...\n",
      "Processing BIOMD0000000935.xml...\n",
      "Processing BIOMD0000000921.xml...\n",
      "Processing BIOMD0000000074.xml...\n",
      "Processing BIOMD0000000712.xml...\n",
      "Processing BIOMD0000000706.xml...\n",
      "Processing BIOMD0000000060.xml...\n",
      "Processing BIOMD0000000048.xml...\n",
      "Processing BIOMD0000000289.xml...\n",
      "Processing BIOMD0000000510.xml...\n",
      "Processing BIOMD0000000276.xml...\n",
      "Processing BIOMD0000000262.xml...\n",
      "Processing BIOMD0000000504.xml...\n",
      "Processing BIOMD0000000538.xml...\n",
      "Processing BIOMD0000000464.xml...\n",
      "Processing BIOMD0000000302.xml...\n",
      "Processing BIOMD0000000316.xml...\n",
      "Processing BIOMD0000001008.xml...\n",
      "Processing BIOMD0000000470.xml...\n",
      "Processing BIOMD0000000458.xml...\n",
      "Processing BIOMD0000001020.xml...\n",
      "Processing BIOMD0000001034.xml...\n",
      "Processing BIOMD0000000869.xml...\n",
      "Processing BIOMD0000000841.xml...\n",
      "Processing BIOMD0000000699.xml...\n",
      "Processing BIOMD0000000855.xml...\n",
      "Processing BIOMD0000000100.xml...\n",
      "Processing BIOMD0000000666.xml...\n",
      "Processing BIOMD0000000672.xml...\n",
      "Processing BIOMD0000000114.xml...\n",
      "Processing BIOMD0000000882.xml...\n",
      "Processing BIOMD0000000896.xml...\n",
      "Processing BIOMD0000000128.xml...\n",
      "Processing BIOMD0000000615.xml...\n",
      "Processing BIOMD0000000173.xml...\n",
      "Processing BIOMD0000000167.xml...\n",
      "Processing BIOMD0000000601.xml...\n",
      "Processing BIOMD0000000629.xml...\n",
      "Processing BIOMD0000000832.xml...\n",
      "Processing BIOMD0000000198.xml...\n",
      "Processing BIOMD0000000826.xml...\n",
      "Processing BIOMD0000000371.xml...\n",
      "Processing BIOMD0000000417.xml...\n",
      "Processing BIOMD0000000403.xml...\n",
      "Processing BIOMD0000000365.xml...\n",
      "Processing BIOMD0000001053.xml...\n",
      "Processing BIOMD0000001047.xml...\n",
      "Processing BIOMD0000000359.xml...\n",
      "Processing BIOMD0000000205.xml...\n",
      "Processing BIOMD0000000563.xml...\n",
      "Processing BIOMD0000000577.xml...\n",
      "Processing BIOMD0000000211.xml...\n",
      "Processing BIOMD0000000239.xml...\n",
      "Processing BIOMD0000000588.xml...\n",
      "Processing BIOMD0000000761.xml...\n",
      "Processing BIOMD0000000007.xml...\n",
      "Processing BIOMD0000000013.xml...\n",
      "Processing BIOMD0000000775.xml...\n",
      "Processing BIOMD0000000985.xml...\n",
      "Processing BIOMD0000000749.xml...\n",
      "Processing BIOMD0000000991.xml...\n",
      "Processing BIOMD0000000946.xml...\n",
      "Processing BIOMD0000000952.xml...\n",
      "Processing BIOMD0000000953.xml...\n",
      "Processing BIOMD0000000947.xml...\n",
      "Processing BIOMD0000000990.xml...\n",
      "Processing BIOMD0000000748.xml...\n",
      "Processing BIOMD0000000984.xml...\n",
      "Processing BIOMD0000000012.xml...\n",
      "Processing BIOMD0000000774.xml...\n",
      "Processing BIOMD0000000760.xml...\n",
      "Processing BIOMD0000000006.xml...\n",
      "Processing BIOMD0000000589.xml...\n",
      "Processing BIOMD0000000238.xml...\n",
      "Processing BIOMD0000000576.xml...\n",
      "Processing BIOMD0000000210.xml...\n",
      "Processing BIOMD0000000204.xml...\n",
      "Processing BIOMD0000000562.xml...\n",
      "Processing BIOMD0000000358.xml...\n",
      "Processing BIOMD0000001046.xml...\n",
      "Processing BIOMD0000001052.xml...\n",
      "Processing BIOMD0000000402.xml...\n",
      "Processing BIOMD0000000364.xml...\n",
      "Processing BIOMD0000000370.xml...\n",
      "Processing BIOMD0000000416.xml...\n",
      "Processing BIOMD0000000199.xml...\n",
      "Processing BIOMD0000000827.xml...\n",
      "Processing BIOMD0000000833.xml...\n",
      "Processing BIOMD0000000628.xml...\n",
      "Processing BIOMD0000000166.xml...\n",
      "Processing BIOMD0000000600.xml...\n",
      "Processing BIOMD0000000614.xml...\n",
      "Processing BIOMD0000000172.xml...\n",
      "Processing BIOMD0000000602.xml...\n",
      "Processing BIOMD0000000164.xml...\n",
      "Processing BIOMD0000000170.xml...\n",
      "Processing BIOMD0000000616.xml...\n",
      "Processing BIOMD0000000158.xml...\n",
      "Processing BIOMD0000000819.xml...\n",
      "Processing BIOMD0000000825.xml...\n",
      "Processing BIOMD0000000831.xml...\n",
      "Processing BIOMD0000001078.xml...\n",
      "Processing BIOMD0000000366.xml...\n",
      "Processing BIOMD0000000400.xml...\n",
      "Processing BIOMD0000000414.xml...\n",
      "Processing BIOMD0000000372.xml...\n",
      "Processing BIOMD0000001044.xml...\n",
      "Processing BIOMD0000000428.xml...\n",
      "Processing BIOMD0000000399.xml...\n",
      "Processing BIOMD0000000212.xml...\n",
      "Processing BIOMD0000000574.xml...\n",
      "Processing BIOMD0000000560.xml...\n",
      "Processing BIOMD0000000206.xml...\n",
      "Processing BIOMD0000000548.xml...\n",
      "Processing BIOMD0000000776.xml...\n",
      "Processing BIOMD0000000010.xml...\n",
      "Processing BIOMD0000000004.xml...\n",
      "Processing BIOMD0000000762.xml...\n",
      "Processing BIOMD0000000038.xml...\n",
      "Processing BIOMD0000000986.xml...\n",
      "Processing BIOMD0000000979.xml...\n",
      "Processing BIOMD0000000951.xml...\n",
      "Processing BIOMD0000000789.xml...\n",
      "Processing BIOMD0000000945.xml...\n",
      "Processing BIOMD0000000944.xml...\n",
      "Processing BIOMD0000000788.xml...\n",
      "Processing BIOMD0000000950.xml...\n",
      "Processing BIOMD0000000978.xml...\n",
      "Processing BIOMD0000000039.xml...\n",
      "Processing BIOMD0000000987.xml...\n",
      "Processing BIOMD0000000005.xml...\n",
      "Processing BIOMD0000000763.xml...\n",
      "Processing BIOMD0000000777.xml...\n",
      "Processing BIOMD0000000011.xml...\n",
      "Processing BIOMD0000000549.xml...\n",
      "Processing BIOMD0000000561.xml...\n",
      "Processing BIOMD0000000207.xml...\n",
      "Processing BIOMD0000000213.xml...\n",
      "Processing BIOMD0000000575.xml...\n",
      "Processing BIOMD0000000398.xml...\n",
      "Processing BIOMD0000000429.xml...\n",
      "Processing BIOMD0000001045.xml...\n",
      "Processing BIOMD0000000415.xml...\n",
      "Processing BIOMD0000000373.xml...\n",
      "Processing BIOMD0000000367.xml...\n",
      "Processing BIOMD0000001079.xml...\n",
      "Processing BIOMD0000000401.xml...\n",
      "Processing BIOMD0000000830.xml...\n",
      "Processing BIOMD0000000824.xml...\n",
      "Processing BIOMD0000000818.xml...\n",
      "Processing BIOMD0000000159.xml...\n",
      "Processing BIOMD0000000171.xml...\n",
      "Processing BIOMD0000000617.xml...\n",
      "Processing BIOMD0000000603.xml...\n",
      "Processing BIOMD0000000165.xml...\n",
      "Processing BIOMD0000000149.xml...\n",
      "Processing BIOMD0000000161.xml...\n",
      "Processing BIOMD0000000607.xml...\n",
      "Processing BIOMD0000000613.xml...\n",
      "Processing BIOMD0000000175.xml...\n",
      "Processing BIOMD0000000820.xml...\n",
      "Processing BIOMD0000000834.xml...\n",
      "Processing BIOMD0000000808.xml...\n",
      "Processing BIOMD0000000439.xml...\n",
      "Processing BIOMD0000001041.xml...\n",
      "Processing BIOMD0000001055.xml...\n",
      "Processing BIOMD0000000405.xml...\n",
      "Processing BIOMD0000000363.xml...\n",
      "Processing BIOMD0000000377.xml...\n",
      "Processing BIOMD0000000411.xml...\n",
      "Processing BIOMD0000000388.xml...\n",
      "Processing BIOMD0000000559.xml...\n",
      "Processing BIOMD0000000571.xml...\n",
      "Processing BIOMD0000000217.xml...\n",
      "Processing BIOMD0000000203.xml...\n",
      "Processing BIOMD0000000565.xml...\n",
      "Processing BIOMD0000000029.xml...\n",
      "Processing BIOMD0000000997.xml...\n",
      "Processing BIOMD0000000983.xml...\n",
      "Processing BIOMD0000000015.xml...\n",
      "Processing BIOMD0000000773.xml...\n",
      "Processing BIOMD0000000767.xml...\n",
      "Processing BIOMD0000000001.xml...\n",
      "Processing BIOMD0000000954.xml...\n",
      "Processing BIOMD0000000940.xml...\n",
      "Processing BIOMD0000000798.xml...\n",
      "Processing BIOMD0000000968.xml...\n",
      "Processing BIOMD0000000969.xml...\n",
      "Processing BIOMD0000000799.xml...\n",
      "Processing BIOMD0000000941.xml...\n",
      "Processing BIOMD0000000955.xml...\n",
      "Processing BIOMD0000000766.xml...\n",
      "Processing BIOMD0000000014.xml...\n",
      "Processing BIOMD0000000772.xml...\n",
      "Processing BIOMD0000000982.xml...\n",
      "Processing BIOMD0000000028.xml...\n",
      "Processing BIOMD0000000996.xml...\n",
      "Processing BIOMD0000000202.xml...\n",
      "Processing BIOMD0000000564.xml...\n",
      "Processing BIOMD0000000570.xml...\n",
      "Processing BIOMD0000000216.xml...\n",
      "Processing BIOMD0000000558.xml...\n",
      "Processing BIOMD0000000389.xml...\n",
      "Processing BIOMD0000000376.xml...\n",
      "Processing BIOMD0000000410.xml...\n",
      "Processing BIOMD0000000404.xml...\n",
      "Processing BIOMD0000000362.xml...\n",
      "Processing BIOMD0000001054.xml...\n",
      "Processing BIOMD0000000438.xml...\n",
      "Processing BIOMD0000001040.xml...\n",
      "Processing BIOMD0000000809.xml...\n",
      "Processing BIOMD0000000835.xml...\n",
      "Processing BIOMD0000000821.xml...\n",
      "Processing BIOMD0000000612.xml...\n",
      "Processing BIOMD0000000174.xml...\n",
      "Processing BIOMD0000000160.xml...\n",
      "Processing BIOMD0000000606.xml...\n",
      "Processing BIOMD0000000148.xml...\n",
      "Processing BIOMD0000000638.xml...\n",
      "Processing BIOMD0000000176.xml...\n",
      "Processing BIOMD0000000610.xml...\n",
      "Processing BIOMD0000000604.xml...\n",
      "Processing BIOMD0000000162.xml...\n",
      "Processing BIOMD0000000189.xml...\n",
      "Processing BIOMD0000000837.xml...\n",
      "Processing BIOMD0000000823.xml...\n",
      "Processing BIOMD0000001056.xml...\n",
      "Processing BIOMD0000000348.xml...\n",
      "Processing BIOMD0000001042.xml...\n",
      "Processing BIOMD0000000412.xml...\n",
      "Processing BIOMD0000000374.xml...\n",
      "Processing BIOMD0000000360.xml...\n",
      "Processing BIOMD0000000406.xml...\n",
      "Processing BIOMD0000000228.xml...\n",
      "Processing BIOMD0000000566.xml...\n",
      "Processing BIOMD0000000200.xml...\n",
      "Processing BIOMD0000000214.xml...\n",
      "Processing BIOMD0000000572.xml...\n",
      "Processing BIOMD0000000599.xml...\n",
      "Processing BIOMD0000000758.xml...\n",
      "Processing BIOMD0000000980.xml...\n",
      "Processing BIOMD0000000994.xml...\n",
      "Processing BIOMD0000000002.xml...\n",
      "Processing BIOMD0000000764.xml...\n",
      "Processing BIOMD0000000770.xml...\n",
      "Processing BIOMD0000000016.xml...\n",
      "Processing BIOMD0000000943.xml...\n",
      "Processing BIOMD0000000957.xml...\n",
      "Processing BIOMD0000000956.xml...\n",
      "Processing BIOMD0000000942.xml...\n",
      "Processing BIOMD0000000771.xml...\n",
      "Processing BIOMD0000000017.xml...\n",
      "Processing BIOMD0000000003.xml...\n",
      "Processing BIOMD0000000765.xml...\n",
      "Processing BIOMD0000000995.xml...\n",
      "Processing BIOMD0000000981.xml...\n",
      "Processing BIOMD0000000759.xml...\n",
      "Processing BIOMD0000000598.xml...\n",
      "Processing BIOMD0000000215.xml...\n",
      "Processing BIOMD0000000573.xml...\n",
      "Processing BIOMD0000000567.xml...\n",
      "Processing BIOMD0000000201.xml...\n",
      "Processing BIOMD0000000229.xml...\n",
      "Processing BIOMD0000001080.xml...\n",
      "Processing BIOMD0000000361.xml...\n",
      "Processing BIOMD0000000407.xml...\n",
      "Processing BIOMD0000000413.xml...\n",
      "Processing BIOMD0000000375.xml...\n",
      "Processing BIOMD0000001043.xml...\n",
      "Processing BIOMD0000000349.xml...\n",
      "Processing BIOMD0000001057.xml...\n",
      "Processing BIOMD0000000822.xml...\n",
      "Processing BIOMD0000000188.xml...\n",
      "Processing BIOMD0000000836.xml...\n",
      "Processing BIOMD0000000605.xml...\n",
      "Processing BIOMD0000000163.xml...\n",
      "Processing BIOMD0000000177.xml...\n",
      "Processing BIOMD0000000611.xml...\n",
      "Processing BIOMD0000000639.xml...\n",
      "Statistics saved to: BioModels_bqbiol_qualifier_statistics_all.csv\n",
      "\n",
      "Model type distribution:\n",
      "  SBML: 1056 models\n",
      "  SBML-qual: 3 models\n",
      "  SBML-fbc: 4 models\n",
      "\n",
      "Discovered ontologies: ['3dmet', 'CHEBI', 'ChEBI', 'GO', 'P24385\"', 'UBERON', 'asap', 'bao', 'bigg.metabolite', 'biocyc', 'bto', 'cas', 'cco', 'chembl.compound', 'cl', 'ec-code', 'eco', 'ecogene', 'efo', 'ensembl', 'envipath', 'fb', 'fma', 'go', 'hgnc.symbol', 'hmdb', 'hp', 'ido', 'inchi_key', 'inchikey', 'insdc', 'interpro', 'kegg', 'kegg.compound', 'kegg.drug', 'kegg.genes', 'kegg.glycan', 'kegg.orthology', 'lipidmaps', 'metanetx.chemical', 'mi', 'mim', 'mirbase', 'mirbase.mature', 'mod', 'mp', 'ncbigi', 'ncbiprotein', 'ncim', 'ncit', 'obi', 'omit', 'opb', 'orphanet', 'pato', 'pirsf', 'pr', 'pubchem.compound', 'pubchem.substance', 'pubmed', 'pw', 'reactome', 'reactome.compound', 'refseq_locus_tag', 'refseq_name', 'refseq_old_locus_tag', 'refseq_synonym', 'sabiork', 'sbo', 'seed.compound', 'sgd', 'slm', 'so', 'taxonomy', 'tcdb', 'uberon']\n",
      "\n",
      "Summary:\n",
      "                         is  isVersionOf  hasVersion  isDescribedBy  hasPart  \\\n",
      "chebi                  3659         1911          65              0     1020   \n",
      "ncbigene               1396           54           0              0        9   \n",
      "uniprot                4215         3689         225              0     3366   \n",
      "pr                       38          215           2              0      284   \n",
      "interpro                 34          246          33              0      345   \n",
      "...                     ...          ...         ...            ...      ...   \n",
      "cas                      14            2           0              0        0   \n",
      "kegg                      0            0           0              0        0   \n",
      "CHEBI                     4            0           0              0        0   \n",
      "P24385\"                   0            1           0              0        0   \n",
      "Models_with_qualifier   454          762         108             14      397   \n",
      "\n",
      "                       isPartOf  hasProperty  isPropertyOf  isEncodedBy  \\\n",
      "chebi                        12          208             1            0   \n",
      "ncbigene                      0            2             0           97   \n",
      "uniprot                      48           59             4            1   \n",
      "pr                            7           67             0            0   \n",
      "interpro                     16            3             0            0   \n",
      "...                         ...          ...           ...          ...   \n",
      "cas                           0            0             0            0   \n",
      "kegg                          0           18             0            0   \n",
      "CHEBI                         0            0             0            0   \n",
      "P24385\"                       0            0             0            0   \n",
      "Models_with_qualifier        54          270            19           37   \n",
      "\n",
      "                       encodes  isHomologTo  occursIn  hasTaxon  isRelatedTo  \n",
      "chebi                        0            3         0         0            0  \n",
      "ncbigene                     0            0         0         0            0  \n",
      "uniprot                    188          273         0         0            0  \n",
      "pr                           1            0         0         0            0  \n",
      "interpro                    10            0         0         0            0  \n",
      "...                        ...          ...       ...       ...          ...  \n",
      "cas                          0            0         0         0            0  \n",
      "kegg                         0            0         0         0            0  \n",
      "CHEBI                        0            0         0         0            0  \n",
      "P24385\"                      0            0         0         0            0  \n",
      "Models_with_qualifier       61           54        63        10            0  \n",
      "\n",
      "[80 rows x 14 columns]\n",
      "\n",
      "Number of models that contain at least one annotation for each qualifier:\n",
      "  is: 454 models\n",
      "  isVersionOf: 762 models\n",
      "  hasVersion: 108 models\n",
      "  isDescribedBy: 14 models\n",
      "  hasPart: 397 models\n",
      "  isPartOf: 54 models\n",
      "  hasProperty: 270 models\n",
      "  isPropertyOf: 19 models\n",
      "  isEncodedBy: 37 models\n",
      "  encodes: 61 models\n",
      "  isHomologTo: 54 models\n",
      "  occursIn: 63 models\n",
      "  hasTaxon: 10 models\n",
      "  isRelatedTo: 0 models\n"
     ]
    }
   ],
   "source": [
    "stats = analyze_bqbiol_qualifier_statistics(\n",
    "    model_dir = \"/Users/luna/Desktop/CRBM/AMAS_proj/Models/BioModels\", \n",
    "    output_file = 'BioModels_bqbiol_qualifier_statistics_all.csv',\n",
    "    verbose = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing 190 models for bqbiol qualifier statistics...\n",
      "Processing 054_PC12-CELL-DIFFERENTIATION_source.sbml...\n",
      "Processing 038_SKBR3-BREAST-CELL-LINE-LONG-TERM_source.sbml...\n",
      "Processing 073_LYMPHOID-AND-MYELOID-CELL-SPECIFICATION_source.sbml...\n",
      "Processing 126_ORF3A_source.sbml...\n",
      "Processing 043_BORDETELLA-BRONCHISEPTICA_source.sbml...\n",
      "Processing 161_DIFFERENTIATION-OF-MONOCYTES_source.sbml...\n",
      "Processing 078_IMMUNE-SYSTEM_source.sbml...\n",
      "Processing 009_YEAST-APOPTOSIS_source.sbml...\n",
      "Processing 072_LYMPHOPOIESIS-REGULATORY-NETWORK_source.sbml...\n",
      "Processing 065_TUMOUR-CELL-INVASION-AND-MIGRATION_source.sbml...\n",
      "Processing 198_PAIR-RULE-MODULE_source.sbml...\n",
      "Processing 090_MAPK-REDUCED-2_source.sbml...\n",
      "Processing 179_MICROENVIRONMENT-CONTROL_source.sbml...\n",
      "Processing 029_TOLL-PATHWAY-OF-DROSOPHILA_source.sbml...\n",
      "Processing 010_CARDIAC-DEVELOPMENT_source.sbml...\n",
      "Processing 172_DROSOPHILA-GAP-D_source.sbml...\n",
      "Processing 193_SENESCENCE-G1S-CHECKPOINT_source.sbml...\n",
      "Processing 123_NSP4-NSP6_source.sbml...\n",
      "Processing 053_PREDICTING-VARIABILITIES-IN-CARDIAC-GENE_source.sbml...\n",
      "Processing 076_SENESCENCE-ASSOCIATED-SECRETORY-PHENOTYPE_source.sbml...\n",
      "Processing 118_INTERFERON-1_source.sbml...\n",
      "Processing 020_APOPTOSIS-NETWORK_source.sbml...\n",
      "Processing 066_CD4-T-CELL-DIFFERENTIATION_source.sbml...\n",
      "Processing 183_ALTERATIONS-IN-BLADDER_source.sbml...\n",
      "Processing 178_MAST-CELL-ACTIVATION_source.sbml...\n",
      "Processing 190_BRAF-TREATMENT-RESPONSE_source.sbml...\n",
      "Processing 218_TCA-CYCLE_source.sbml...\n",
      "Processing 047_FGF-PATHWAY-OF-DROSOPHILA_source.sbml...\n",
      "Processing 169_DROSOPHILA-GAP-A_source.sbml...\n",
      "Processing 077_SIGNALLING-PATHWAY-FOR-BUTANOL-PRODUCTION_source.sbml...\n",
      "Processing 141_HIGH-OSMOLARITY-AND-MATING-PATHWAYS_source.sbml...\n",
      "Processing 145_MELANOGENESIS_source.sbml...\n",
      "Processing 104_DROSOPHILA-CELL-CYCLE_source.sbml...\n",
      "Processing 184_P53-MDM2-NETWORK_source.sbml...\n",
      "Processing 018_EGFR-ERBB-SIGNALING_source.sbml...\n",
      "Processing 176_MYELOFIBROTIC-MICROENVIRONMENT_source.sbml...\n",
      "Processing 151_TCR-REDOX-METABOLISM_source.sbml...\n",
      "Processing 132_VIRUS-REPLICATION-CYCLE_source.sbml...\n",
      "Processing 046_B-BRONCHISEPTICA-AND-T-RETORTAEFORMIS_source.sbml...\n",
      "Processing 016_IL-1-SIGNALING_source.sbml...\n",
      "Processing 036_HCC1954-BREAST-CELL-LINE-SHORT-TERM_source.sbml...\n",
      "Processing 040_T-CELL-DIFFERENTIATION_source.sbml...\n",
      "Processing 191_SEGMENT-POLARITY-1-CELL_source.sbml...\n",
      "Processing 122_NSP14_source.sbml...\n",
      "Processing 200_LUNG-CANCER-CELL-CYCLE_source.sbml...\n",
      "Processing 156_CONTROL-OF-TH1-TH2-TH17-TREG-DIFFERENTATION-REDUCED_source.sbml...\n",
      "Processing 034_HCC1954-BREAST-CELL-LINE-LONG-TERM_source.sbml...\n",
      "Processing 022_B-CELL-DIFFERENTIATION_source.sbml...\n",
      "Processing 035_BT474-BREAST-CELL-LINE-SHORT-TERM_source.sbml...\n",
      "Processing 082_TCR-TLR5-SIGNALING-2018_source.sbml...\n",
      "Processing 007_CORTICAL-AREA-DEVELOPMENT_source.sbml...\n",
      "Processing 189_TRP-BIOSYNTHESIS_source.sbml...\n",
      "Processing 070_MAPK-CANCER-CELL-FATE_source.sbml...\n",
      "Processing 155_CONTROL-OF-TH1-TH2-TH17-TREG-DIFFERENTATION_source.sbml...\n",
      "Processing 061_TUMOR-MICROENVIRONMENT-IN-LYMPHOBLASTIC-LEUKEAMIA_source.sbml...\n",
      "Processing 139_ACUTE-RESPONSES-DURING-HYPERINSULINEMIA_source.sbml...\n",
      "Processing 069_IRON-ACQUISITION-AND-STRESS-RESPONSE_source.sbml...\n",
      "Processing 106_STOMATAL-RESTING-STATE_source.sbml...\n",
      "Processing 159_BUDDING-YEAST-CORE_source.sbml...\n",
      "Processing 095_FISSION-YEAST-2008_source.sbml...\n",
      "Processing 175_SEA-URCHIN_source.sbml...\n",
      "Processing 063_LAC-OPERON_source.sbml...\n",
      "Processing 013_CHOLESTEROL-REGULATORY-PATHWAY_source.sbml...\n",
      "Processing 032_T-CELL-SIGNALLING-2006_source.sbml...\n",
      "Processing 097_DROSOPHILA-WINGS-AP_source.sbml...\n",
      "Processing 214_FOXO3-PATHWAY_source.sbml...\n",
      "Processing 165_EGGSHELL-PATTERNING-PHENOMOENOLOGICAL_source.sbml...\n",
      "Processing 187_MAMMAL-SEX-DETERMINATION-1-CELL_source.sbml...\n",
      "Processing 136_EGF-TNF-ALPHA-SIGNALLING-PATHWAY_source.sbml...\n",
      "Processing 131_TGFB-PATHWAY_source.sbml...\n",
      "Processing 015_NEUROTRANSMITTER-SIGNALING-PATHWAY_source.sbml...\n",
      "Processing 216_PPARGC1A-PATHWAY_source.sbml...\n",
      "Processing 219_WNT-PI3K-AKT-PATHWAY_source.sbml...\n",
      "Processing 135_SIGNAL-TRANSDUCTION_source.sbml...\n",
      "Processing 203_DDR-SIGNALLING-PATHWAYS_source.sbml...\n",
      "Processing 088_MIR-9-NEUROGENESIS_source.sbml...\n",
      "Processing 108_GEROCONVERSION_source.sbml...\n",
      "Processing 196_T-LYMPHOCYTE-SPECIFICATION_source.sbml...\n",
      "Processing 003_MAMMALIAN-CELL-CYCLE_source.sbml...\n",
      "Processing 062_CD4-T-CELL-DIFFERENTIATION-AND-PLASTICITY_source.sbml...\n",
      "Processing 042_TOL-REGULATORY-NETWORK_source.sbml...\n",
      "Processing 099_YEAST-HYPHAL-TRANSITION_source.sbml...\n",
      "Processing 120_KYNURENINE-PATHWAY_source.sbml...\n",
      "Processing 071_CASTRATION-RESISTANT-PROSTATE-CANCER_source.sbml...\n",
      "Processing 005_FA-BRCA-PATHWAY_source.sbml...\n",
      "Processing 021_BODY-SEGMENTATION-IN-DROSOPHILA-2013_source.sbml...\n",
      "Processing 164_EGGSHELL-PATTERNING-MECHANISTIC_source.sbml...\n",
      "Processing 195_CTLA4-PD1-CHECKPOINT-INHIBITORS_source.sbml...\n",
      "Processing 019_IL-6-SIGNALING_source.sbml...\n",
      "Processing 166_DROSOPHILA-JAK-STAT-PATHWAY_source.sbml...\n",
      "Processing 059_BORTEZOMIB-RESPONSES-IN-MYELOMA-CELLS_source.sbml...\n",
      "Processing 181_MULTILEVEL-CELL-CYCLE_source.sbml...\n",
      "Processing 008_DEATH-RECEPTOR-SIGNALING_source.sbml...\n",
      "Processing 130_RENIN-ANGIOTENSIN_source.sbml...\n",
      "Processing 148_AGS-CELL-FATE-DECISION_source.sbml...\n",
      "Processing 109_ASYMMETRIC-CELL-DIVISION-A_source.sbml...\n",
      "Processing 014_T-LGL-SURVIVAL-NETWORK-2008_source.sbml...\n",
      "Processing 068_AURORA-KINASE-A-IN-NEUROBLASTOMA_source.sbml...\n",
      "Processing 170_DROSOPHILA-GAP-B_source.sbml...\n",
      "Processing 134_RHEUMATOID-ARTHRITIS_source.sbml...\n",
      "Processing 152_TCR-REDOX-METABOLISM-REDUCED_source.sbml...\n",
      "Processing 186_CHICKEN-SEX-DETERMINATION-REDUCED_source.sbml...\n",
      "Processing 004_ERBB-RECEPTOR-SIGNALING_source.sbml...\n",
      "Processing 086_TUMOUR-CELL-INVASION-AND-MIGRATION-REDUCED_source.sbml...\n",
      "Processing 157_CONTROL-OF-TH-DIFFERENTATION_source.sbml...\n",
      "Processing 188_MAMMAL-SEX-DETERMINATION-2-CELL_source.sbml...\n",
      "Processing 064_METABOLIC-INTERACTIONS-IN-GUT-MICROBIOME_source.sbml...\n",
      "Processing 057_FANCONI-ANEMIA-AND-CHECKPOINT-RECOVERY_source.sbml...\n",
      "Processing 044_TRICHOSTRONGYLUS-RETORTAEFORMIS_source.sbml...\n",
      "Processing 031_CELL-CYCLE-TRANSCRIPTION_source.sbml...\n",
      "Processing 167_MESODERM-SPECIFICATION-IN-DROSOPHILA_source.sbml...\n",
      "Processing 194_VULVAR-PRECURSOR-CELLS_source.sbml...\n",
      "Processing 185_CHICKEN-SEX-DETERMINATION_source.sbml...\n",
      "Processing 163_DROSOPHILA-EGF-PATHWAY_source.sbml...\n",
      "Processing 142_BLOOD-STEM-CELL_source.sbml...\n",
      "Processing 149_AGS-CELL-FATE-DECISION-REDUCED_source.sbml...\n",
      "Processing 058_ARABIDOPSIS-THALIANA-CELL-CYCLE_source.sbml...\n",
      "Processing 143_BREAST-CANCER-INHIBITORS_source.sbml...\n",
      "Processing 146_BUDDING-YEAST-FAURE-2009_source.sbml...\n",
      "Processing 133_ROOT-STEM-CELL-2010_source.sbml...\n",
      "Processing 055_HUMAN-GONADAL-SEX-DETERMINATION_source.sbml...\n",
      "Processing 213_DOPAMINE-TRANSCRIPTION_source.sbml...\n",
      "Processing 085_REPROGRAMMING-TESTES-DERIVED-STEM-CELLS_source.sbml...\n",
      "Processing 096_ERBB-REGULATED-G1-S-TRANSITION_source.sbml...\n",
      "Processing 111_APOPTOSIS_source.sbml...\n",
      "Processing 017_DIFFERENTIATION-OF-T-LYMPHOCYTES_source.sbml...\n",
      "Processing 037_SKBR3-BREAST-CELL-LINE-SHORT-TERM_source.sbml...\n",
      "Processing 110_ASYMMETRIC-CELL-DIVISION-B_source.sbml...\n",
      "Processing 039_HIV-1-INTERACTIONS-WITH-T-CELL-SIGNALING_source.sbml...\n",
      "Processing 116_HMOX1-PATHWAY_source.sbml...\n",
      "Processing 006_HGF-SIGNALING-IN-KERATINOCYTES_source.sbml...\n",
      "Processing 162_DROSOPHILA-DPP-PATHWAY_source.sbml...\n",
      "Processing 045_HH-PATHWAY-OF-DROSOPHILA_source.sbml...\n",
      "Processing 081_TLR5-SIGNALING-2018_source.sbml...\n",
      "Processing 049_OXIDATIVE-STRESS-PATHWAY_source.sbml...\n",
      "Processing 147_BUDDING-YEAST-EXIT-MODULE_source.sbml...\n",
      "Processing 177_LYMPHOID-CELL-SPECIFICATION_source.sbml...\n",
      "Processing 150_CELL-FATE-DECISION-MULTISCALE_source.sbml...\n",
      "Processing 125_ORF10-CUL2-PATHWAY_source.sbml...\n",
      "Processing 128_PYRIMIDINE-DEPRIVATION_source.sbml...\n",
      "Processing 113_ER-STRESS_source.sbml...\n",
      "Processing 023_MAMMALIAN-CELL-CYCLE-2006_source.sbml...\n",
      "Processing 115_E-PROTEIN_source.sbml...\n",
      "Processing 025_T-LGL-SURVIVAL-NETWORK-2011_source.sbml...\n",
      "Processing 137_SIGNALLING-IN-LIVER-CANCER_source.sbml...\n",
      "Processing 093_IMMUNE-CHECKPOINT-INHIBITORS_source.sbml...\n",
      "Processing 041_INFLUENZA-VIRUS-REPLICATION-CYCLE_source.sbml...\n",
      "Processing 119_JNK-PATHWAY_source.sbml...\n",
      "Processing 067_REGULATION-OF-L-ARABINOSE-OPERON_source.sbml...\n",
      "Processing 083_SIGNALING-IN-PROSTATE-CANCER_source.sbml...\n",
      "Processing 153_CONTROL-OF-PROLIFERATION_source.sbml...\n",
      "Processing 182_BOOLEAN-CELL-CYCLE_source.sbml...\n",
      "Processing 117_IFN-LAMBDA_source.sbml...\n",
      "Processing 089_MAPK-REDUCED-1_source.sbml...\n",
      "Processing 091_MAPK-REDUCED-3_source.sbml...\n",
      "Processing 060_STOMATAL-OPENING_source.sbml...\n",
      "Processing 028_VEGF-PATHWAY-OF-DROSOPHILA_source.sbml...\n",
      "Processing 048_GLUCOSE-REPRESSION-SIGNALING-2009_source.sbml...\n",
      "Processing 197_ANTERIOR-POSTERIOR-BOUNDARY_source.sbml...\n",
      "Processing 080_TCR-SIGNALING-2018_source.sbml...\n",
      "Processing 171_DROSOPHILA-GAP-C_source.sbml...\n",
      "Processing 050_CD4-T-CELL-SIGNALING_source.sbml...\n",
      "Processing 158_LAMBDA-PHAGE-LYSOGENY_source.sbml...\n",
      "Processing 204_HUMAN-BRAIN-ORGANOIDS_source.sbml...\n",
      "Processing 112_COAGULATION-PATHWAY_source.sbml...\n",
      "Processing 124_NSP9-PROTEIN_source.sbml...\n",
      "Processing 154_CONTROL-OF-TH1-TH2-DIFFERENTIATION_source.sbml...\n",
      "Processing 168_DROSOPHILA-NOTCH-PATHWAY_source.sbml...\n",
      "Processing 012_T-CELL-RECEPTOR-SIGNALING_source.sbml...\n",
      "Processing 052_SEPTATION-INITIATION-NETWORK_source.sbml...\n",
      "Processing 215_MTOR-PATHWAY_source.sbml...\n",
      "Processing 026_BUDDING-YEAST-CELL-CYCLE-2009_source.sbml...\n",
      "Processing 217_PRKN-PATHWAY_source.sbml...\n",
      "Processing 114_ETC_source.sbml...\n",
      "Processing 002_SIGNAL-TRANSDUCTION-IN-FIBROBLASTS_source.sbml...\n",
      "Processing 127_PAMP-SIGNALING_source.sbml...\n",
      "Processing 180_MORPHOGENETIC-CHECKPOINT_source.sbml...\n",
      "Processing 051_COLITIS-ASSOCIATED-COLON-CANCER_source.sbml...\n",
      "Processing 160_IL17-DIFFERENTIAL-EXPRESSION_source.sbml...\n",
      "Processing 001_SIGNALING-IN-MACROPHAGE-ACTIVATION_source.sbml...\n",
      "Processing 056_IGVH-MUTATIONS-IN-LYMPHOCYTIC-LEUKEMIA_source.sbml...\n",
      "Processing 033_BT474-BREAST-CELL-LINE-LONG-TERM_source.sbml...\n",
      "Processing 129_RTC-AND-TRANSCRIPTION_source.sbml...\n",
      "Processing 192_SEGMENT-POLARITY-6-CELL_source.sbml...\n",
      "Processing 030_SPZ-NETWORK-OF-DROSOPHILA_source.sbml...\n",
      "Processing 024_BUDDING-YEAST-CELL-CYCLE_source.sbml...\n",
      "Processing 074_T-LGL-SURVIVAL-NETWORK-2011-REDUCED_source.sbml...\n",
      "Processing 027_WG-PATHWAY-OF-DROSOPHILA_source.sbml...\n",
      "Processing 011_GUARD-CELL-ABSCISIC-ACID-SIGNALING_source.sbml...\n",
      "Processing 121_NLRP3-ACTIVATION_source.sbml...\n",
      "Statistics saved to: BioDivine_bqbiol_qualifier_statistics_all.csv\n",
      "\n",
      "Model type distribution:\n",
      "  SBML: 0 models\n",
      "  SBML-qual: 190 models\n",
      "  SBML-fbc: 0 models\n",
      "\n",
      "Discovered ontologies: ['brenda', 'cas', 'chembl.compound', 'chemspider', 'doi', 'drugbank', 'ec-code', 'ensembl', 'hgnc', 'hgnc.symbol', 'hmdb', 'intact', 'interpro', 'kegg.compound', 'mesh', 'ncbiprotein', 'obo.chebi', 'obo.go', 'omim', 'pdb', 'pubchem.compound', 'pubchem.substance', 'pubmed', 'reactome', 'refseq', 'taxonomy', 'vmhmetabolite', 'wikipathways', 'wikipedia.en']\n",
      "\n",
      "Summary:\n",
      "                       is  isVersionOf  hasVersion  isDescribedBy  hasPart  \\\n",
      "chebi                   0            0           0              0        0   \n",
      "ncbigene                9            1           0            377        0   \n",
      "uniprot                80           12           0            385        1   \n",
      "pubmed                 15            0           0           2770        0   \n",
      "hgnc                    2            2           0            174       16   \n",
      "ncbiprotein            18            0           0              7        7   \n",
      "taxonomy                2            0           0             41        0   \n",
      "doi                     2            0           0             18        0   \n",
      "obo.go                 37            1           8             30        1   \n",
      "obo.chebi              77           13           0             58        1   \n",
      "wikipathways            1            0           0              3        0   \n",
      "mesh                    1            0           0             33        9   \n",
      "reactome                8            1          25              0        0   \n",
      "vmhmetabolite          17            0           0              1        0   \n",
      "refseq                  3            0           0             46        0   \n",
      "ensembl                 2            0           0             42        0   \n",
      "hgnc.symbol             3            0           0             77       16   \n",
      "wikipedia.en           11            0           0              5        0   \n",
      "pubchem.compound       45            0           0              6        0   \n",
      "cas                    10            0           0              4        0   \n",
      "kegg.compound          15            0           0             10        0   \n",
      "chemspider             14            0           0              4        0   \n",
      "hmdb                   10            0           0              7        0   \n",
      "chembl.compound         3            0           0              0        0   \n",
      "interpro                0            0           0              0        1   \n",
      "pubchem.substance       1            0           0              0        0   \n",
      "ec-code                 0            0           0              1        0   \n",
      "brenda                  0            1           0              0        0   \n",
      "drugbank                0            0           0              1        0   \n",
      "intact                  0            0           0              2        0   \n",
      "pdb                     0            0           0              1        0   \n",
      "omim                    0            0           0              0        0   \n",
      "Models_with_qualifier  17            4           7             78        4   \n",
      "\n",
      "                       isPartOf  hasProperty  isPropertyOf  isEncodedBy  \\\n",
      "chebi                         0            0             0            0   \n",
      "ncbigene                      0            0             0           16   \n",
      "uniprot                       0            0             0          180   \n",
      "pubmed                        0            0             0            0   \n",
      "hgnc                          0            0             0          105   \n",
      "ncbiprotein                   0            0             0           18   \n",
      "taxonomy                      0            0             0           35   \n",
      "doi                           0            0             0            0   \n",
      "obo.go                        0            0             0           69   \n",
      "obo.chebi                     0            0             0          276   \n",
      "wikipathways                  0            0             0            1   \n",
      "mesh                          0            0             0            7   \n",
      "reactome                      0            0             0            9   \n",
      "vmhmetabolite                 0            0             0           33   \n",
      "refseq                        0            0             0            3   \n",
      "ensembl                       0            0             0            0   \n",
      "hgnc.symbol                   0            0             0          101   \n",
      "wikipedia.en                  0            0             0            0   \n",
      "pubchem.compound              0            0             0            4   \n",
      "cas                           0            0             0            0   \n",
      "kegg.compound                 0            0             0            1   \n",
      "chemspider                    0            0             0            0   \n",
      "hmdb                          0            0             0            0   \n",
      "chembl.compound               0            0             0            0   \n",
      "interpro                      4            0             0            6   \n",
      "pubchem.substance             0            0             0            1   \n",
      "ec-code                       0            0             0            0   \n",
      "brenda                        0            0             0            0   \n",
      "drugbank                      0            0             0            0   \n",
      "intact                        0            0             0            0   \n",
      "pdb                           0            0             0            0   \n",
      "omim                          0            0             0            1   \n",
      "Models_with_qualifier         3            0             0           23   \n",
      "\n",
      "                       encodes  isHomologTo  occursIn  hasTaxon  isRelatedTo  \n",
      "chebi                        0            0         0         0            0  \n",
      "ncbigene                     1            0         0         0            0  \n",
      "uniprot                      0            0         0         0            0  \n",
      "pubmed                       0            0         0         0            0  \n",
      "hgnc                         0            0         0         0            0  \n",
      "ncbiprotein                  0            0         0         0            0  \n",
      "taxonomy                     0            0         0         0            0  \n",
      "doi                          0            0         1         0            0  \n",
      "obo.go                       0            0         0         0            0  \n",
      "obo.chebi                    0            0         0         0            0  \n",
      "wikipathways                 0            0         0         0            0  \n",
      "mesh                         0            0         0         0            0  \n",
      "reactome                     0            0         0         0            0  \n",
      "vmhmetabolite                0            0         0         0            0  \n",
      "refseq                       0            0         0         0            0  \n",
      "ensembl                      0            0         0         0            0  \n",
      "hgnc.symbol                 11            0         0         0            0  \n",
      "wikipedia.en                 0            0         0         0            0  \n",
      "pubchem.compound             0            0         0         0            0  \n",
      "cas                          0            0         0         0            0  \n",
      "kegg.compound                0            0         0         0            0  \n",
      "chemspider                   0            0         0         0            0  \n",
      "hmdb                         0            0         0         0            0  \n",
      "chembl.compound              0            0         0         0            0  \n",
      "interpro                     0            0         0         0            0  \n",
      "pubchem.substance            0            0         0         0            0  \n",
      "ec-code                      0            0         0         0            0  \n",
      "brenda                       0            0         0         0            0  \n",
      "drugbank                     0            0         0         0            0  \n",
      "intact                       0            0         0         0            0  \n",
      "pdb                          0            0         0         0            0  \n",
      "omim                         0            0         0         0            0  \n",
      "Models_with_qualifier        2            0         1         0            0  \n",
      "\n",
      "Number of models that contain at least one annotation for each qualifier:\n",
      "  is: 17 models\n",
      "  isVersionOf: 4 models\n",
      "  hasVersion: 7 models\n",
      "  isDescribedBy: 78 models\n",
      "  hasPart: 4 models\n",
      "  isPartOf: 3 models\n",
      "  hasProperty: 0 models\n",
      "  isPropertyOf: 0 models\n",
      "  isEncodedBy: 23 models\n",
      "  encodes: 2 models\n",
      "  isHomologTo: 0 models\n",
      "  occursIn: 1 models\n",
      "  hasTaxon: 0 models\n",
      "  isRelatedTo: 0 models\n"
     ]
    }
   ],
   "source": [
    "stats = analyze_bqbiol_qualifier_statistics(\n",
    "    model_dir = \"/Users/luna/Desktop/CRBM/AMAS_proj/Models/BioDivine\", \n",
    "    output_file = 'BioDivine_bqbiol_qualifier_statistics_all.csv',\n",
    "    verbose = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO Filtered results saved to: results/biomd_chebi_rag_llama-4_top3_prompt_adjusted_hasPart.csv\n"
     ]
    }
   ],
   "source": [
    "filtered = filter_qualifiers_in_results(\n",
    "    results_csv = 'results/biomd_chebi_rag_llama-4_top3_prompt_adjusted.csv', \n",
    "    bqbiol_qualifiers = ['hasPart'],\n",
    "    model_dir = \"/Users/luna/Desktop/CRBM/AMAS_proj/Models/BioModels\",\n",
    "    output_csv = \"results/biomd_chebi_rag_llama-4_top3_prompt_adjusted_hasPart.csv\",\n",
    "    entity_type = \"chemical\",\n",
    "    database = \"chebi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Showing all results\n",
      "Number of models assessed: 95\n",
      "Number of models with predictions: 86\n",
      "Number of annotations evaluated: 1043\n",
      "Average accuracy (per model): 0.41\n",
      "Ave. recall (formula): 0.38\n",
      "Ave. precision (formula): 0.31\n",
      "Ave. recall (exact): 0.32\n",
      "Ave. precision (exact): 0.13\n",
      "Average accuracy (per species): 0.26\n",
      "Ave. recall (formula, per species): 0.24\n",
      "Ave. precision (formula, per species): 0.18\n",
      "Ave. recall (exact, per species): 0.21\n",
      "Ave. precision (exact, per species): 0.08\n",
      "Ave. total time (per model): 11.58\n",
      "Ave. total time (per element, per model): 1.05\n",
      "Ave. LLM time (per model): 11.29\n",
      "Ave. LLM time (per element, per model): 1.03\n",
      "Average number of predictions per species: 2.60\n"
     ]
    }
   ],
   "source": [
    "print_evaluation_results(\"results/biomd_chebi_rag_llama-4_top3_prompt_adjusted_hasPart.csv\",ref_results_csv = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Showing all results\n",
      "Number of models assessed: 95\n",
      "Number of models with predictions: 88\n",
      "Number of annotations evaluated: 1043\n",
      "Average accuracy (per model): 0.38\n",
      "Ave. recall (formula): 0.36\n",
      "Ave. precision (formula): 0.27\n",
      "Ave. recall (exact): 0.31\n",
      "Ave. precision (exact): 0.12\n",
      "Average accuracy (per species): 0.19\n",
      "Ave. recall (formula, per species): 0.18\n",
      "Ave. precision (formula, per species): 0.12\n",
      "Ave. recall (exact, per species): 0.14\n",
      "Ave. precision (exact, per species): 0.06\n",
      "Ave. total time (per model): 5.09\n",
      "Ave. total time (per element, per model): 0.46\n",
      "Ave. LLM time (per model): 4.74\n",
      "Ave. LLM time (per element, per model): 0.43\n",
      "Average number of predictions per species: 2.47\n"
     ]
    }
   ],
   "source": [
    "print_evaluation_results(\"results/biomd_chebi_rag_meta-llama_top3_prompt_adjusted_hasPart.csv\",ref_results_csv = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO Filtered results saved to: results/biomd_chebi_rag_meta-llama_top3_prompt_adjusted_is+isVersionOf.csv\n"
     ]
    }
   ],
   "source": [
    "filtered = filter_qualifiers_in_results(\n",
    "    results_csv = \"results/biomd_chebi_rag_meta-llama_top3_prompt_adjusted.csv\", \n",
    "    bqbiol_qualifiers = ['is','isVersionOf'],\n",
    "    model_dir = \"/Users/luna/Desktop/CRBM/AMAS_proj/Models/BioModels\",\n",
    "    output_csv = \"results/biomd_chebi_rag_meta-llama_top3_prompt_adjusted_is+isVersionOf.csv\",\n",
    "    entity_type = \"chemical\",\n",
    "    database = \"chebi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Showing all results\n",
      "Number of models assessed: 317\n",
      "Number of models with predictions: 313\n",
      "Number of annotations evaluated: 9654\n",
      "Average accuracy (per model): 0.86\n",
      "Ave. recall (formula): 0.86\n",
      "Ave. precision (formula): 0.65\n",
      "Ave. recall (exact): 0.73\n",
      "Ave. precision (exact): 0.25\n",
      "Average accuracy (per species): 0.40\n",
      "Ave. recall (formula, per species): 0.40\n",
      "Ave. precision (formula, per species): 0.29\n",
      "Ave. recall (exact, per species): 0.29\n",
      "Ave. precision (exact, per species): 0.12\n",
      "Ave. total time (per model): 5.03\n",
      "Ave. total time (per element, per model): 0.17\n",
      "Ave. LLM time (per model): 4.51\n",
      "Ave. LLM time (per element, per model): 0.15\n",
      "Average number of predictions per species: 2.87\n"
     ]
    }
   ],
   "source": [
    "print_evaluation_results(\"results/biomd_chebi_rag_meta-llama_top3_prompt_adjusted_is+isVersionOf.csv\",ref_results_csv = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Showing all results\n",
      "Number of models assessed: 190\n",
      "Number of models with predictions: 187\n",
      "Number of annotations evaluated: 3480\n",
      "Average accuracy (per model): 0.88\n",
      "Ave. recall (formula): 0.87\n",
      "Ave. precision (formula): 0.66\n",
      "Ave. recall (exact): 0.76\n",
      "Ave. precision (exact): 0.26\n",
      "Average accuracy (per species): 0.59\n",
      "Ave. recall (formula, per species): 0.59\n",
      "Ave. precision (formula, per species): 0.43\n",
      "Ave. recall (exact, per species): 0.49\n",
      "Ave. precision (exact, per species): 0.17\n",
      "Ave. total time (per model): 5.29\n",
      "Ave. total time (per element, per model): 0.29\n",
      "Ave. LLM time (per model): 4.84\n",
      "Ave. LLM time (per element, per model): 0.26\n",
      "Average number of predictions per species: 2.90\n"
     ]
    }
   ],
   "source": [
    "print_evaluation_results(\"results/biomd_chebi_rag_meta-llama_top3_prompt_adjusted_filter_is.csv\",ref_results_csv = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered = filter_qualifiers_in_results(\n",
    "    results_csv = \"results/biomd_chebi_rag_meta-llama_top3_prompt_adjusted.csv\", \n",
    "    bqbiol_qualifiers = ['is'],\n",
    "    model_dir = \"/Users/luna/Desktop/CRBM/AMAS_proj/Models/BioModels\",\n",
    "    output_csv = \"results/biomd_chebi_rag_meta-llama_top3_prompt_adjusted_filter_is.csv\",\n",
    "    entity_type = \"chemical\",\n",
    "    database = \"chebi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for all qualifiers\n",
      "Showing all results\n",
      "Number of models assessed: 340\n",
      "Number of models with predictions: 338\n",
      "Number of annotations evaluated: 10771\n",
      "Average accuracy (per model): 0.79\n",
      "Ave. recall (formula): 0.78\n",
      "Ave. precision (formula): 0.59\n",
      "Ave. recall (exact): 0.68\n",
      "Ave. precision (exact): 0.24\n",
      "Average accuracy (per species): 0.40\n",
      "Ave. recall (formula, per species): 0.39\n",
      "Ave. precision (formula, per species): 0.28\n",
      "Ave. recall (exact, per species): 0.28\n",
      "Ave. precision (exact, per species): 0.13\n",
      "Ave. total time (per model): 10.16\n",
      "Ave. total time (per element, per model): 0.32\n",
      "Ave. LLM time (per model): 9.77\n",
      "Ave. LLM time (per element, per model): 0.31\n",
      "Average number of predictions per species: 2.84\n",
      "Results for is qualifier\n",
      "Showing all results\n",
      "Number of models assessed: 190\n",
      "Number of models with predictions: 189\n",
      "Number of annotations evaluated: 3480\n",
      "Average accuracy (per model): 0.88\n",
      "Ave. recall (formula): 0.88\n",
      "Ave. precision (formula): 0.66\n",
      "Ave. recall (exact): 0.78\n",
      "Ave. precision (exact): 0.27\n",
      "Average accuracy (per species): 0.57\n",
      "Ave. recall (formula, per species): 0.57\n",
      "Ave. precision (formula, per species): 0.42\n",
      "Ave. recall (exact, per species): 0.48\n",
      "Ave. precision (exact, per species): 0.17\n",
      "Ave. total time (per model): 11.71\n",
      "Ave. total time (per element, per model): 0.64\n",
      "Ave. LLM time (per model): 11.34\n",
      "Ave. LLM time (per element, per model): 0.62\n",
      "Average number of predictions per species: 2.90\n"
     ]
    }
   ],
   "source": [
    "print(\"Results for all qualifiers\")\n",
    "print_evaluation_results(\"results/biomd_chebi_rag_llama-4_top3_prompt_adjusted.csv\",ref_results_csv = None)\n",
    "print(\"Results for is qualifier\")\n",
    "print_evaluation_results(\"results/biomd_chebi_rag_llama-4_top3_prompt_adjusted_filter_is.csv\",ref_results_csv = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Failure reason"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts of each unique value in 'Wrong_reason':\n",
      "Wrong_reason\n",
      "NaN    10721\n",
      "6.0       29\n",
      "5.0        6\n",
      "3.0        5\n",
      "2.0        4\n",
      "1.0        3\n",
      "8.0        2\n",
      "4.0        1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "excel_file = \"/Users/luna/Desktop/CRBM/AMAS_proj/AAAIM/tests/results/biomd_chebi_rag_llama-4_top10_prompt_adjusted_review.xlsx\"\n",
    "df = pd.read_excel(excel_file)\n",
    "# Count occurrences of each unique value in 'Wrong_reason' column\n",
    "reason_counts = df['Wrong_reason'].value_counts(dropna=False)\n",
    "print(\"Counts of each unique value in 'Wrong_reason':\")\n",
    "print(reason_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run for complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Test model found: /Users/luna/Desktop/CRBM/AMAS_proj/Models/BioModels/BIOMD0000000088.xml\n"
     ]
    }
   ],
   "source": [
    "test_model_file = \"/Users/luna/Desktop/CRBM/AMAS_proj/Models/BioModels/BIOMD0000000088.xml\"\n",
    "if os.path.exists(test_model_file):\n",
    "    print(f\" Test model found: {test_model_file}\")\n",
    "else:\n",
    "    print(f\" Test model not found: {test_model_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'s187': ['15996'],\n",
       " 's173': ['16595'],\n",
       " 's214': ['15996'],\n",
       " 's231': ['17552'],\n",
       " 's252': ['15996'],\n",
       " 's311': ['18035'],\n",
       " 's436': ['15996'],\n",
       " 's438': ['17552'],\n",
       " 's439': ['15996'],\n",
       " 's440': ['17552'],\n",
       " 's443': ['15996'],\n",
       " 's449': ['15996'],\n",
       " 's456': ['15996'],\n",
       " 's506': ['29108'],\n",
       " 's512': ['29108'],\n",
       " 's513': ['29108'],\n",
       " 's520': ['29108'],\n",
       " 's526': ['29108'],\n",
       " 's539': ['29108'],\n",
       " 's546': ['29108'],\n",
       " 's551': ['29108'],\n",
       " 's564': ['29108', '15996'],\n",
       " 's565': ['29108', '18035'],\n",
       " 's566': ['29108'],\n",
       " 's567': ['29108', '15996'],\n",
       " 's568': ['29108', '18348'],\n",
       " 's569': ['29108']}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_species_with_chebi_annotations(test_model_file,['hasPart'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GO']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "identifiers_pattern = r'http[s]?://identifiers\\.org/([^/:]+)'\n",
    "text = \"http://identifiers.org/GO:0042752\"\n",
    "re.findall(identifiers_pattern, text)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aaaim",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
